<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <link href='https://fonts.googleapis.com/css?family=Chivo:900' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/pygment_trac.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <title>Jannovar by charite</title>
  </head>

  <body>
    <div id="container">
      <div class="inner">

        <header>
          <h1>Jannovar</h1>
          <h2>Java framework for annotating VCF files with gene/transcript-based annotations</h2>
        </header>

        <section id="downloads" class="clearfix">
          <a href="https://github.com/charite/jannovar/zipball/master" id="download-zip" class="button"><span>Download .zip</span></a>
          <a href="https://github.com/charite/jannovar/tarball/master" id="download-tar-gz" class="button"><span>Download .tar.gz</span></a>
          <a href="https://github.com/charite/jannovar" id="view-on-github" class="button"><span>View on GitHub</span></a>
        </section>

        <hr>

        <section id="main_content">
          <h1>
<a id="jannovar-a-java-library-for-exome-annotation" class="anchor" href="#jannovar-a-java-library-for-exome-annotation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Jannovar: A Java library for Exome/Genome Annotation</h1>

<p>Transcript-based annotation and pedigree analysis are two basic steps in the computational analysis of whole-exome/genome sequencing experiments in disease-gene discovery projects or diagnostics. Jannovar is a stand-alone Java application as well as a Java library designed to be used in larger software frameworks for exome/genome analysis. Jannovar uses an interval tree to identify all transcripts affected by a given variant, and provides HGVS-compliant annotations for both for variants affecting coding sequences and splice junctions as well as UTR sequences and non-coding RNA transcripts. Jannovar can also perform family-based pedigree analysis with VCF files with data from members of a family segregating a Mendelian disorder. Using a desktop computer, Jannovar requires a few seconds to annotate a typical VCF file with exome/genome data.</p>

<h2>
<a id="downloads" class="anchor" href="#downloads" aria-hidden="true"><span class="octicon octicon-link"></span></a>Using Jannovar</h2>

<p>Jannovar can be used from the command line to annotate VCF files, but it is also a complete Java programming library that can be used within larger applications.
For instance, the <a href="http://www.sanger.ac.uk/resources/databases/exomiser/">Exomiser</a> is a Web server that can be used to prioritise candidate genes in exome sequencing studies. 
<a href="http://compbio.charite.de/PhenIX">PhenIX</a> is a Web server that provides a ranked list of
candidate genes in diagnostic exome or gene panel sequencing. <a href="https://phenomecentral.org/">Phenome Central</a> is used for genomic matchmaking. </p>
        
<h2> Jannovar official manual</h2>

          <p>For usage information, installation and other help please visit our <a href="http://jannovar.readthedocs.io/en/master/">Jannovar Manual</a>. Report bugs on GitHub <a href="https://github.com/charite/jannovar/issues">issue tracker</a>
<p>If you are reporting a bug, please include:</p>
 <ul>
  <li>Your operating system name and version.</li>
  <li>Any details about your local setup that might be helpful in troubleshooting.</li>
  <li>Detailed steps to reproduce the bug.</li>
</ul>
          
    
<h2>
<a id="downloading-the-source-code" class="anchor" href="#downloading-the-source-code" aria-hidden="true"><span class="octicon octicon-link"></span></a>Tutorial (old version)</h2>

<p>A complete  <a href="tutorial/doc/JannovarTutorial.pdf">tutorial</a> is available that explains how to use Jannovar both as a command-line application and as a programming library.
A compressed  <a href="tutorial/tutorial.tgz">archive</a> of source code examples and other files is provided for the tutorial.</p>

<h2>
<a id="license" class="anchor" href="#license" aria-hidden="true"><span class="octicon octicon-link"></span></a>License</h2>

<p>Jannovar is distributed under a BSD2 license.</p>

<h2>
<a id="publication" class="anchor" href="#publication" aria-hidden="true"><span class="octicon octicon-link"></span></a>Publication</h2>

<p>JÃ¤ger, Marten, et al. <a href="http://onlinelibrary.wiley.com/doi/10.1002/humu.22531/full">"Jannovar: A Java Library for Exome Annotation."</a> Human mutation 35.5 (2014): 548-555.</p>
        </section>

        <footer>
          Jannovar is maintained by <a href="https://github.com/charite">charite</a><br>
          This page was generated by <a href="http://pages.github.com">GitHub Pages</a>. Tactile theme by <a href="https://twitter.com/jasonlong">Jason Long</a>.
        </footer>

        
      </div>
    </div>
  </body>
</html>
