<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (1.8.0_101) on Wed Aug 17 17:05:57 CEST 2016 -->
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<title>VariantEffect (Jannovar 0.17 API)</title>
<meta name="date" content="2016-08-17">
<link rel="stylesheet" type="text/css" href="../../../../../stylesheet.css" title="Style">
<script type="text/javascript" src="../../../../../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="VariantEffect (Jannovar 0.17 API)";
        }
    }
    catch(err) {
    }
//-->
var methods = {"i0":10,"i1":10,"i2":10,"i3":10,"i4":10,"i5":10,"i6":10,"i7":10,"i8":10,"i9":10,"i10":9,"i11":9};
var tabs = {65535:["t0","All Methods"],1:["t1","Static Methods"],2:["t2","Instance Methods"],8:["t4","Concrete Methods"]};
var altColor = "altColor";
var rowColor = "rowColor";
var tableTab = "tableTab";
var activeTableTab = "activeTableTab";
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/VariantEffect.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../index-all.html">Index</a></li>
</ul>
<div class="aboutLanguage">Jannovar, 0.17</div>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantAnnotator.html" title="class in de.charite.compbio.jannovar.annotation"><span class="typeNameLink">Prev&nbsp;Class</span></a></li>
<li>Next&nbsp;Class</li>
</ul>
<ul class="navList">
<li><a href="../../../../../index.html?de/charite/compbio/jannovar/annotation/VariantEffect.html" target="_top">Frames</a></li>
<li><a href="VariantEffect.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../../../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#enum.constant.summary">Enum Constants</a>&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li><a href="#enum.constant.detail">Enum Constants</a>&nbsp;|&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="subTitle">de.charite.compbio.jannovar.annotation</div>
<h2 title="Enum VariantEffect" class="title">Enum VariantEffect</h2>
</div>
<div class="contentContainer">
<ul class="inheritance">
<li><a href="http://docs.oracle.com/javase/8/docs/api/java/lang/Object.html?is-external=true" title="class or interface in java.lang">java.lang.Object</a></li>
<li>
<ul class="inheritance">
<li><a href="http://docs.oracle.com/javase/8/docs/api/java/lang/Enum.html?is-external=true" title="class or interface in java.lang">java.lang.Enum</a>&lt;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a>&gt;</li>
<li>
<ul class="inheritance">
<li>de.charite.compbio.jannovar.annotation.VariantEffect</li>
</ul>
</li>
</ul>
</li>
</ul>
<div class="description">
<ul class="blockList">
<li class="blockList">
<dl>
<dt>All Implemented Interfaces:</dt>
<dd><a href="http://docs.oracle.com/javase/8/docs/api/java/io/Serializable.html?is-external=true" title="class or interface in java.io">Serializable</a>, <a href="http://docs.oracle.com/javase/8/docs/api/java/lang/Comparable.html?is-external=true" title="class or interface in java.lang">Comparable</a>&lt;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a>&gt;</dd>
</dl>
<hr>
<br>
<pre>public enum <span class="typeNameLabel">VariantEffect</span>
extends <a href="http://docs.oracle.com/javase/8/docs/api/java/lang/Enum.html?is-external=true" title="class or interface in java.lang">Enum</a>&lt;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a>&gt;</pre>
<div class="block">These codes reflect the possible types of variants that we call for an exome.

 The values in this enum are given in the putative order of impact (more severe to less severe). The documentation
 gives the sequence ontology (SO) ID and the SO description. Also, the documentation of each value explains whether
 Jannovar generates this annotation or not.</div>
<dl>
<dt><span class="simpleTagLabel">Author:</span></dt>
<dd><a href="mailto:peter.robinson@charite.de">Peter Robinson</a>, <a href="mailto:marten.jaeger@charite.de">Marten Jaeger</a>, <a href="mailto:manuel.holtgrewe@charite.de">Manuel Holtgrewe</a>, <a href="mailto:max.schubach@charite.de">Max Schubach</a></dd>
</dl>
</li>
</ul>
</div>
<div class="summary">
<ul class="blockList">
<li class="blockList">
<!-- =========== ENUM CONSTANT SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="enum.constant.summary">
<!--   -->
</a>
<h3>Enum Constant Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Enum Constant Summary table, listing enum constants, and an explanation">
<caption><span>Enum Constants</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Enum Constant and Description</th>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#Z:Z_SMALLEST_HIGH_IMPACT">_SMALLEST_HIGH_IMPACT</a></span></code>
<div class="block">Marker for smallest <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation"><code>VariantEffect</code></a> with <a href="../../../../../de/charite/compbio/jannovar/annotation/PutativeImpact.html#HIGH"><code>PutativeImpact.HIGH</code></a> impact.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#Z:Z_SMALLEST_LOW_IMPACT">_SMALLEST_LOW_IMPACT</a></span></code>
<div class="block">Marker for smallest <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation"><code>VariantEffect</code></a> with <a href="../../../../../de/charite/compbio/jannovar/annotation/PutativeImpact.html#LOW"><code>PutativeImpact.LOW</code></a> impact.</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#Z:Z_SMALLEST_MODERATE_IMPACT">_SMALLEST_MODERATE_IMPACT</a></span></code>
<div class="block">Marker for smallest <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation"><code>VariantEffect</code></a> with <a href="../../../../../de/charite/compbio/jannovar/annotation/PutativeImpact.html#MODERATE"><code>PutativeImpact.MODERATE</code></a> impact.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#CHROMOSOME">CHROMOSOME</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0000340">SO:0000340</a> Structural unit
 composed of a nucleic acid molecule which controls its own replication through the interaction of specific
 proteins at one or more origins of replication (is a: replicon).</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#CHROMOSOME_NUMBER_VARIATION">CHROMOSOME_NUMBER_VARIATION</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:1000182">SO:1000182</a> A kind of chromosome
 variation where the chromosome complement is not an exact multiple of the haploid number (is a
 chromosome_variation).</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#CODING_SEQUENCE_VARIANT">CODING_SEQUENCE_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001580">SO:0001580</a> A sequence variant
 that changes the coding sequence (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#CODING_TRANSCRIPT_VARIANT"><code>CODING_TRANSCRIPT_VARIANT</code></a>, <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#EXON_VARIANT"><code>EXON_VARIANT</code></a>).</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#CODING_TRANSCRIPT_INTRON_VARIANT">CODING_TRANSCRIPT_INTRON_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001969">SO:0001969</a> A sequence variant
 that changes non-coding intro sequence in a non-coding transcript (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#CODING_TRANSCRIPT_VARIANT"><code>CODING_TRANSCRIPT_VARIANT</code></a>,
 <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#INTRON_VARIANT"><code>INTRON_VARIANT</code></a>).</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#CODING_TRANSCRIPT_VARIANT">CODING_TRANSCRIPT_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001968">SO:0001968</a> A transcript variant
 of a protein coding gene (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#TRANSCRIPT_VARIANT"><code>TRANSCRIPT_VARIANT</code></a>).</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#COMPLEX_SUBSTITUTION">COMPLEX_SUBSTITUTION</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:1000005">SO:1000005</a> When no simple or
 well defined DNA mutation event describes the observed DNA change, the keyword "complex" should be used.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#CONSERVED_INTERGENIC_VARIANT">CONSERVED_INTERGENIC_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002017">SO:0002017</a> A sequence variant
 located in a conserved intergenic region, between genes (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#INTERGENIC_VARIANT"><code>INTERGENIC_VARIANT</code></a>).</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#CONSERVED_INTRON_VARIANT">CONSERVED_INTRON_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002018">SO:0002018</a> A transcript variant
 occurring within a conserved region of an intron (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#INTRON_VARIANT"><code>INTRON_VARIANT</code></a>).</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#CUSTOM">CUSTOM</a></span></code>
<div class="block">Variant in a user-specified custom region.</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#DIRECT_TANDEM_DUPLICATION">DIRECT_TANDEM_DUPLICATION</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:1000039">SO:1000039</a> A tandem duplication
 where the individual regions are in the same orientation (is a: tandem_duplication).</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#DISRUPTIVE_INFRAME_DELETION">DISRUPTIVE_INFRAME_DELETION</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001826">SO:0001826</a> An inframe decrease
 in cds length that deletes bases from the coding sequence starting within an existing codon (is a:
 <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#INFRAME_DELETION"><code>INFRAME_DELETION</code></a>).</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#DISRUPTIVE_INFRAME_INSERTION">DISRUPTIVE_INFRAME_INSERTION</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001824">SO:0001824</a> An inframe increase
 in cds length that inserts one or more codons into the coding sequence within an existing codon (is a:
 <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#INFRAME_INSERTION"><code>INFRAME_INSERTION</code></a>).</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#DOWNSTREAM_GENE_VARIANT">DOWNSTREAM_GENE_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001632">SO:0001632</a> A sequence variant
 located 3' of a gene (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#INTERGENIC_VARIANT"><code>INTERGENIC_VARIANT</code></a>).</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#EXON_LOSS_VARIANT">EXON_LOSS_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001572">SO:0001572</a> A sequence variant
 whereby an exon is lost from the transcript (is a (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#SPLICING_VARIANT"><code>SPLICING_VARIANT</code></a>), <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#TRANSCRIPT_VARIANT"><code>TRANSCRIPT_VARIANT</code></a>
 ).</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#EXON_VARIANT">EXON_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001791">SO:0001791</a> A sequence variant
 that changes exon sequence (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#TRANSCRIPT_VARIANT"><code>TRANSCRIPT_VARIANT</code></a>).</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#FEATURE_TRUNCATION">FEATURE_TRUNCATION</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001906">SO:0001906</a> A sequence variant
 that causes the reduction of a genomic feature, with regard to the reference sequence (is a: feature_variant).</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#FIVE_PRIME_UTR_EXON_VARIANT">FIVE_PRIME_UTR_EXON_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002092">SO:0002092</a> A UTR variant of the
 5' UTR (is a: 5_prime_UTR_variant; is a: UTR_variant).</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#FIVE_PRIME_UTR_INTRON_VARIANT">FIVE_PRIME_UTR_INTRON_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002091">SO:0002091</a> A UTR variant
 between 5' UTRs (is a: 5_prime_UTR_variant; is a: UTR_variant).</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#FIVE_PRIME_UTR_PREMATURE_START_CODON_GAIN_VARIANT">FIVE_PRIME_UTR_PREMATURE_START_CODON_GAIN_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001983">SO:0001983</a> A 5' UTR variant
 where a premature start codon is introduced, moved or lost (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#FIVE_PRIME_UTR_EXON_VARIANT"><code>FIVE_PRIME_UTR_EXON_VARIANT</code></a> or
 <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#FIVE_PRIME_UTR_INTRON_VARIANT"><code>FIVE_PRIME_UTR_INTRON_VARIANT</code></a>).</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#FIVE_PRIME_UTR_TRUNCATION">FIVE_PRIME_UTR_TRUNCATION</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002013">SO:0002013</a> A sequence variant
 that causes the reduction of a the 5'UTR with regard to the reference sequence (is a:
 <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#FIVE_PRIME_UTR_EXON_VARIANT"><code>FIVE_PRIME_UTR_EXON_VARIANT</code></a> or <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#FIVE_PRIME_UTR_INTRON_VARIANT"><code>FIVE_PRIME_UTR_INTRON_VARIANT</code></a>)

 Jannovar does <b>not</b> yield use this at the moment.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#FRAMESHIFT_ELONGATION">FRAMESHIFT_ELONGATION</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001909">SO:0001909</a> A frameshift variant
 that causes the translational reading frame to be extended relative to the reference feature (is a
 <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#FRAMESHIFT_VARIANT"><code>FRAMESHIFT_VARIANT</code></a>, internal_feature_elongation).</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#FRAMESHIFT_TRUNCATION">FRAMESHIFT_TRUNCATION</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001910">SO:0001910</a> A frameshift variant
 that causes the translational reading frame to be shortened relative to the reference feature (is a
 <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#FRAMESHIFT_VARIANT"><code>FRAMESHIFT_VARIANT</code></a>, internal_feature_truncation).</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#FRAMESHIFT_VARIANT">FRAMESHIFT_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001589">SO:0001589</a>A sequence variant
 which causes a disruption of the translational reading frame, because the number of nucleotides inserted or
 deleted is not a multiple of threee (is a: protein_altering_variant).</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#GENE_VARIANT">GENE_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001564">SO:0001564</a> A sequence variant
 where the structure of the gene is changed (is a: feature_variant).</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#INFRAME_DELETION">INFRAME_DELETION</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001822">SO:0001822</a> An inframe non
 synonymous variant that deletes bases from the coding sequence (is a: inframe_indel, feature_truncation).</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#INFRAME_INSERTION">INFRAME_INSERTION</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001821">SO:0001821</a> An inframe non
 synonymous variant that inserts bases into in the coding sequence (is a: inframe_indel,
 internal_feature_elongation).</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#INITIATOR_CODON_VARIANT">INITIATOR_CODON_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001582">SO:0001582</a> A codon variant that
 changes at least one base of the first codon of a transcript (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#CODING_SEQUENCE_VARIANT"><code>CODING_SEQUENCE_VARIANT</code></a>, children:
 start_retained_variant, start_lost).</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#INTERGENIC_REGION">INTERGENIC_REGION</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0000605">SO:0000605</a> A region containing
 or overlapping no genes that is bounded on either side by a gene, or bounded by a gene and the end of the
 chromosome (is a: biological_region).</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#INTERGENIC_VARIANT">INTERGENIC_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001628">SO:0001628</a> A sequence variant
 located in the intergenic region, between genes (is a: feature_variant).</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#INTERNAL_FEATURE_ELONGATION">INTERNAL_FEATURE_ELONGATION</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001908">SO:0001908</a> A sequence variant
 that causes the extension of a genomic feature from within the feature rather than from the terminus of the
 feature, with regard to the reference sequence.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#INTRAGENIC_VARIANT">INTRAGENIC_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002011">SO:0002011</a> A variant that
 occurs within a gene but falls outside of all transcript features.</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#INTRON_VARIANT">INTRON_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001627">SO:0001627</a> A transcript variant
 occurring within an intron (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#TRANSCRIPT_VARIANT"><code>TRANSCRIPT_VARIANT</code></a>).</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#MIRNA">MIRNA</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0000276">SO:0000276</a> Variant affects a
 miRNA (is a: miRNA_primary_transcript, small_regulatory_ncRNA).</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#MISSENSE_VARIANT">MISSENSE_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001583">SO:0001583</a> A sequence variant,
 that changes one or more bases, resulting in a different amino acid sequence but where the length is preserved.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#MNV">MNV</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002007">SO:0002007</a> An MNV is a multiple
 nucleotide variant (substitution) in which the inserted sequence is the same length as the replaced sequence (is
 a: substitution).</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#NON_CODING_TRANSCRIPT_EXON_VARIANT">NON_CODING_TRANSCRIPT_EXON_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001792">SO:0001792</a> A sequence variant
 that changes non-coding exon sequence in a non-coding transcript (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#NON_CODING_TRANSCRIPT_VARIANT"><code>NON_CODING_TRANSCRIPT_VARIANT</code></a>,
 <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#EXON_VARIANT"><code>EXON_VARIANT</code></a>).</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#NON_CODING_TRANSCRIPT_INTRON_VARIANT">NON_CODING_TRANSCRIPT_INTRON_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001970">SO:0001970</a> A sequence variant
 that changes non-coding intro sequence in a non-coding transcript (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#NON_CODING_TRANSCRIPT_VARIANT"><code>NON_CODING_TRANSCRIPT_VARIANT</code></a>,
 <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#INTRON_VARIANT"><code>INTRON_VARIANT</code></a>).</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#NON_CODING_TRANSCRIPT_VARIANT">NON_CODING_TRANSCRIPT_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001619">SO:0001619</a> (is a:
 <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#TRANSCRIPT_VARIANT"><code>TRANSCRIPT_VARIANT</code></a>).</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#RARE_AMINO_ACID_VARIANT">RARE_AMINO_ACID_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002008">SO:0002008</a> A sequence variant
 whereby at least one base of a codon encoding a rare amino acid is changed, resulting in a different encoded
 amino acid (children: selenocysteine_loss, pyrrolysine_loss).</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#REGULATORY_REGION_VARIANT">REGULATORY_REGION_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001566">SO:0001566</a> A sequence variant
 located within a regulatory region (is a: feature_variant).</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#SEQUENCE_VARIANT">SEQUENCE_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001060">SO:0001060</a> Top level term for
 variants, can be used for marking "uknown effect".</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#SPLICE_ACCEPTOR_VARIANT">SPLICE_ACCEPTOR_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001574">SO:0001574</a> A splice variant
 that changes the 2 base region at the 3' end of an intron (is a <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#SPLICE_REGION_VARIANT"><code>SPLICE_REGION_VARIANT</code></a>).</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#SPLICE_DONOR_VARIANT">SPLICE_DONOR_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001575">SO:0001575</a> A splice variant
 that changes the 2 base pair region at the 5' end of an intron (is a <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#SPLICE_REGION_VARIANT"><code>SPLICE_REGION_VARIANT</code></a>).</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#SPLICE_REGION_VARIANT">SPLICE_REGION_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001630">SO:0001630</a> A sequence variant
 in which a change has occurred within the region of the splice site, either within 1-3 bases of the exon or 3-8
 bases of the intron (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#SPLICING_VARIANT"><code>SPLICING_VARIANT</code></a>).</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#SPLICING_VARIANT">SPLICING_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001568">SO:0001568</a> A sequence variant
 that changes the process of splicing (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#GENE_VARIANT"><code>GENE_VARIANT</code></a>).</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#START_LOST">START_LOST</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002012">SO:0002012</a> A codon variant that
 changes at least one base of the canonical start codon (is a: initiator_codon_variant).</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#STOP_GAINED">STOP_GAINED</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001587">SO:0001587</a> A sequence variant
 whereby at least one base of a codon is changed, resulting in a premature stop codon, leading to a shortened
 transcript (is a: nonsynonymous_variant, feature_truncation).</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#STOP_LOST">STOP_LOST</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001578">SO:0001578</a> A sequence variant
 where at least one base of the terminator codon (stop) is changed, resulting in an elongated transcript (is a:
 nonsynonymous variant, terminator_codon_variant, feature_elongation)</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#STOP_RETAINED_VARIANT">STOP_RETAINED_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001567">SO:0001567</a> A sequence variant
 where at least one base in the terminator codon is changed, but the terminator remains (is a:
 <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#SYNONYMOUS_VARIANT"><code>SYNONYMOUS_VARIANT</code></a>, terminator_codon_variant).</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#STRUCTURAL_VARIANT">STRUCTURAL_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001537">SO:0001537</a> A sequence variant
 that changes one or more sequence features (is a: sequence variant).</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#SYNONYMOUS_VARIANT">SYNONYMOUS_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001819">SO:0001819</a> A sequence variant
 where there is no resulting change to the encoded amino acid (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#CODING_SEQUENCE_VARIANT"><code>CODING_SEQUENCE_VARIANT</code></a>, children:
 start_retained_variant, stop_retained_variant).</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#TF_BINDING_SITE_VARIANT">TF_BINDING_SITE_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001782">SO:0001782</a> A sequence variant
 located within a transcription factor binding site (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#REGULATORY_REGION_VARIANT"><code>REGULATORY_REGION_VARIANT</code></a>).</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#THREE_PRIME_UTR_EXON_VARIANT">THREE_PRIME_UTR_EXON_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002089">SO:0002089</a> A UTR variant of the
 3' UTR (is a: 3_prime_UTR_variant; is a: UTR_variant).</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#THREE_PRIME_UTR_INTRON_VARIANT">THREE_PRIME_UTR_INTRON_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002090">SO:0002090</a> A UTR variant
 between 3' UTRs (is a: 3_prime_UTR_variant; is a: UTR_variant).</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#THREE_PRIME_UTR_TRUNCATION">THREE_PRIME_UTR_TRUNCATION</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002015">SO:0002015</a> A sequence variant
 that causes the reduction of a the 3' UTR with regard to the reference sequence (is a:
 <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#FIVE_PRIME_UTR_EXON_VARIANT"><code>FIVE_PRIME_UTR_EXON_VARIANT</code></a> or <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#FIVE_PRIME_UTR_INTRON_VARIANT"><code>FIVE_PRIME_UTR_INTRON_VARIANT</code></a>).</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#TRANSCRIPT_ABLATION">TRANSCRIPT_ABLATION</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001893">SO:0001893</a> A feature ablation
 whereby the deleted region includes a transcript feature (is a: feature_ablation)

 <b>Not</b> used in Jannovar annotations.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#TRANSCRIPT_VARIANT">TRANSCRIPT_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001576">SO:0001576</a> A sequence variant
 that changes the structure of the transcript (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#GENE_VARIANT"><code>GENE_VARIANT</code></a>).</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#UPSTREAM_GENE_VARIANT">UPSTREAM_GENE_VARIANT</a></span></code>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001631">SO:0001631</a> A sequence variant
 located 5' of a gene (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#INTERGENIC_VARIANT"><code>INTERGENIC_VARIANT</code></a>).</div>
</td>
</tr>
</table>
</li>
</ul>
<!-- =========== FIELD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="field.summary">
<!--   -->
</a>
<h3>Field Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Field Summary table, listing fields, and an explanation">
<caption><span>Fields</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Field and Description</th>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static com.google.common.base.Predicate&lt;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a>&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#IS_INTRONIC">IS_INTRONIC</a></span></code>
<div class="block"><code>Predicate</code> for testing whether a <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation"><code>VariantEffect</code></a> is intronic.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static com.google.common.base.Predicate&lt;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a>&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#IS_SPLICING">IS_SPLICING</a></span></code>
<div class="block"><code>Predicate</code> for testing whether a <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation"><code>VariantEffect</code></a> is related to splicing.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static com.google.common.base.Function&lt;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a>,<a href="http://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#TO_LEGACY_NAME">TO_LEGACY_NAME</a></span></code>
<div class="block"><code>Function</code> for converting from <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation"><code>VariantEffect</code></a> to legacy term.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static com.google.common.base.Function&lt;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a>,<a href="http://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#TO_SO_TERM">TO_SO_TERM</a></span></code>
<div class="block"><code>Function</code> for converting from <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation"><code>VariantEffect</code></a> to SO term String.</div>
</td>
</tr>
</table>
</li>
</ul>
<!-- ========== METHOD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="method.summary">
<!--   -->
</a>
<h3>Method Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Method Summary table, listing methods, and an explanation">
<caption><span id="t0" class="activeTableTab"><span>All Methods</span><span class="tabEnd">&nbsp;</span></span><span id="t1" class="tableTab"><span><a href="javascript:show(1);">Static Methods</a></span><span class="tabEnd">&nbsp;</span></span><span id="t2" class="tableTab"><span><a href="javascript:show(2);">Instance Methods</a></span><span class="tabEnd">&nbsp;</span></span><span id="t4" class="tableTab"><span><a href="javascript:show(8);">Concrete Methods</a></span><span class="tabEnd">&nbsp;</span></span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tr id="i0" class="altColor">
<td class="colFirst"><code><a href="../../../../../de/charite/compbio/jannovar/annotation/PutativeImpact.html" title="enum in de.charite.compbio.jannovar.annotation">PutativeImpact</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#getImpact--">getImpact</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i1" class="rowColor">
<td class="colFirst"><code><a href="http://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#getLegacyTerm--">getLegacyTerm</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i2" class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#getNumber--">getNumber</a></span>()</code>
<div class="block">Forward to <code>ordinal()</code> member function.</div>
</td>
</tr>
<tr id="i3" class="rowColor">
<td class="colFirst"><code><a href="http://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#getSequenceOID--">getSequenceOID</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i4" class="altColor">
<td class="colFirst"><code><a href="http://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#getSequenceOntologyTerm--">getSequenceOntologyTerm</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i5" class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#isIntronic--">isIntronic</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i6" class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#isOffExome--">isOffExome</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i7" class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#isOffTranscript--">isOffTranscript</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i8" class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#isSplicing--">isSplicing</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i9" class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#isStructural--">isStructural</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i10" class="altColor">
<td class="colFirst"><code>static <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#valueOf-java.lang.String-">valueOf</a></span>(<a href="http://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;name)</code>
<div class="block">Returns the enum constant of this type with the specified name.</div>
</td>
</tr>
<tr id="i11" class="rowColor">
<td class="colFirst"><code>static <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a>[]</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#values--">values</a></span>()</code>
<div class="block">Returns an array containing the constants of this enum type, in
the order they are declared.</div>
</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.java.lang.Enum">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;java.lang.<a href="http://docs.oracle.com/javase/8/docs/api/java/lang/Enum.html?is-external=true" title="class or interface in java.lang">Enum</a></h3>
<code><a href="http://docs.oracle.com/javase/8/docs/api/java/lang/Enum.html?is-external=true#compareTo-E-" title="class or interface in java.lang">compareTo</a>, <a href="http://docs.oracle.com/javase/8/docs/api/java/lang/Enum.html?is-external=true#equals-java.lang.Object-" title="class or interface in java.lang">equals</a>, <a href="http://docs.oracle.com/javase/8/docs/api/java/lang/Enum.html?is-external=true#getDeclaringClass--" title="class or interface in java.lang">getDeclaringClass</a>, <a href="http://docs.oracle.com/javase/8/docs/api/java/lang/Enum.html?is-external=true#hashCode--" title="class or interface in java.lang">hashCode</a>, <a href="http://docs.oracle.com/javase/8/docs/api/java/lang/Enum.html?is-external=true#name--" title="class or interface in java.lang">name</a>, <a href="http://docs.oracle.com/javase/8/docs/api/java/lang/Enum.html?is-external=true#ordinal--" title="class or interface in java.lang">ordinal</a>, <a href="http://docs.oracle.com/javase/8/docs/api/java/lang/Enum.html?is-external=true#toString--" title="class or interface in java.lang">toString</a>, <a href="http://docs.oracle.com/javase/8/docs/api/java/lang/Enum.html?is-external=true#valueOf-java.lang.Class-java.lang.String-" title="class or interface in java.lang">valueOf</a></code></li>
</ul>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.java.lang.Object">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;java.lang.<a href="http://docs.oracle.com/javase/8/docs/api/java/lang/Object.html?is-external=true" title="class or interface in java.lang">Object</a></h3>
<code><a href="http://docs.oracle.com/javase/8/docs/api/java/lang/Object.html?is-external=true#getClass--" title="class or interface in java.lang">getClass</a>, <a href="http://docs.oracle.com/javase/8/docs/api/java/lang/Object.html?is-external=true#notify--" title="class or interface in java.lang">notify</a>, <a href="http://docs.oracle.com/javase/8/docs/api/java/lang/Object.html?is-external=true#notifyAll--" title="class or interface in java.lang">notifyAll</a>, <a href="http://docs.oracle.com/javase/8/docs/api/java/lang/Object.html?is-external=true#wait--" title="class or interface in java.lang">wait</a>, <a href="http://docs.oracle.com/javase/8/docs/api/java/lang/Object.html?is-external=true#wait-long-" title="class or interface in java.lang">wait</a>, <a href="http://docs.oracle.com/javase/8/docs/api/java/lang/Object.html?is-external=true#wait-long-int-" title="class or interface in java.lang">wait</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="details">
<ul class="blockList">
<li class="blockList">
<!-- ============ ENUM CONSTANT DETAIL =========== -->
<ul class="blockList">
<li class="blockList"><a name="enum.constant.detail">
<!--   -->
</a>
<h3>Enum Constant Detail</h3>
<a name="CHROMOSOME_NUMBER_VARIATION">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>CHROMOSOME_NUMBER_VARIATION</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> CHROMOSOME_NUMBER_VARIATION</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:1000182">SO:1000182</a> A kind of chromosome
 variation where the chromosome complement is not an exact multiple of the haploid number (is a
 chromosome_variation).

 <b>Not</b> used in Jannovar annotations.</div>
</li>
</ul>
<a name="TRANSCRIPT_ABLATION">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>TRANSCRIPT_ABLATION</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> TRANSCRIPT_ABLATION</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001893">SO:0001893</a> A feature ablation
 whereby the deleted region includes a transcript feature (is a: feature_ablation)

 <b>Not</b> used in Jannovar annotations.</div>
</li>
</ul>
<a name="EXON_LOSS_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>EXON_LOSS_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> EXON_LOSS_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001572">SO:0001572</a> A sequence variant
 whereby an exon is lost from the transcript (is a (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#SPLICING_VARIANT"><code>SPLICING_VARIANT</code></a>), <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#TRANSCRIPT_VARIANT"><code>TRANSCRIPT_VARIANT</code></a>
 ).

 <b>Not</b> used in Jannovar annotations.</div>
</li>
</ul>
<a name="FRAMESHIFT_ELONGATION">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>FRAMESHIFT_ELONGATION</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> FRAMESHIFT_ELONGATION</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001909">SO:0001909</a> A frameshift variant
 that causes the translational reading frame to be extended relative to the reference feature (is a
 <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#FRAMESHIFT_VARIANT"><code>FRAMESHIFT_VARIANT</code></a>, internal_feature_elongation).</div>
</li>
</ul>
<a name="FRAMESHIFT_TRUNCATION">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>FRAMESHIFT_TRUNCATION</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> FRAMESHIFT_TRUNCATION</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001910">SO:0001910</a> A frameshift variant
 that causes the translational reading frame to be shortened relative to the reference feature (is a
 <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#FRAMESHIFT_VARIANT"><code>FRAMESHIFT_VARIANT</code></a>, internal_feature_truncation).</div>
</li>
</ul>
<a name="FRAMESHIFT_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>FRAMESHIFT_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> FRAMESHIFT_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001589">SO:0001589</a>A sequence variant
 which causes a disruption of the translational reading frame, because the number of nucleotides inserted or
 deleted is not a multiple of threee (is a: protein_altering_variant).

 Used for frameshift variant for the case where there is no stop codon any more and the rare case in which the
 transcript length is retained.</div>
</li>
</ul>
<a name="INTERNAL_FEATURE_ELONGATION">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>INTERNAL_FEATURE_ELONGATION</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> INTERNAL_FEATURE_ELONGATION</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001908">SO:0001908</a> A sequence variant
 that causes the extension of a genomic feature from within the feature rather than from the terminus of the
 feature, with regard to the reference sequence.

 In Jannovar, used to annotate a <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#COMPLEX_SUBSTITUTION"><code>COMPLEX_SUBSTITUTION</code></a> that does not lead to a frameshift and increases
 the transcript length.</div>
</li>
</ul>
<a name="FEATURE_TRUNCATION">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>FEATURE_TRUNCATION</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> FEATURE_TRUNCATION</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001906">SO:0001906</a> A sequence variant
 that causes the reduction of a genomic feature, with regard to the reference sequence (is a: feature_variant).

 The term <a href="http://www.sequenceontology.org/browser/current_svn/term/INTERNAL_FEATURE_TRUNCATION">
 INTERNAL_FEATURE_TRUNCATION</a> would be more fitting but is not available in SO.

 In Jannovar, used to annotate a <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#COMPLEX_SUBSTITUTION"><code>COMPLEX_SUBSTITUTION</code></a> that does not lead to a frameshift and decreases
 the transcript length.</div>
</li>
</ul>
<a name="MNV">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>MNV</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> MNV</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002007">SO:0002007</a> An MNV is a multiple
 nucleotide variant (substitution) in which the inserted sequence is the same length as the replaced sequence (is
 a: substitution).</div>
</li>
</ul>
<a name="COMPLEX_SUBSTITUTION">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>COMPLEX_SUBSTITUTION</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> COMPLEX_SUBSTITUTION</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:1000005">SO:1000005</a> When no simple or
 well defined DNA mutation event describes the observed DNA change, the keyword "complex" should be used. Usually
 there are multiple equally plausible explanations for the change (is a: substitution).

 Used together with <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#INTERNAL_FEATURE_ELONGATION"><code>INTERNAL_FEATURE_ELONGATION</code></a> or <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#FEATURE_TRUNCATION"><code>FEATURE_TRUNCATION</code></a> to describe an variant
 that does not lead to a frameshift but a changed transcript length. Used together with
 <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#FRAMESHIFT_ELONGATION"><code>FRAMESHIFT_ELONGATION</code></a> or <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#FRAMESHIFT_TRUNCATION"><code>FRAMESHIFT_TRUNCATION</code></a> if the substitution leads to a frameshift
 variant.</div>
</li>
</ul>
<a name="STOP_GAINED">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>STOP_GAINED</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> STOP_GAINED</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001587">SO:0001587</a> A sequence variant
 whereby at least one base of a codon is changed, resulting in a premature stop codon, leading to a shortened
 transcript (is a: nonsynonymous_variant, feature_truncation).</div>
</li>
</ul>
<a name="STOP_LOST">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>STOP_LOST</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> STOP_LOST</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001578">SO:0001578</a> A sequence variant
 where at least one base of the terminator codon (stop) is changed, resulting in an elongated transcript (is a:
 nonsynonymous variant, terminator_codon_variant, feature_elongation)</div>
</li>
</ul>
<a name="START_LOST">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>START_LOST</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> START_LOST</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002012">SO:0002012</a> A codon variant that
 changes at least one base of the canonical start codon (is a: initiator_codon_variant).</div>
</li>
</ul>
<a name="SPLICE_ACCEPTOR_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>SPLICE_ACCEPTOR_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> SPLICE_ACCEPTOR_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001574">SO:0001574</a> A splice variant
 that changes the 2 base region at the 3' end of an intron (is a <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#SPLICE_REGION_VARIANT"><code>SPLICE_REGION_VARIANT</code></a>).</div>
</li>
</ul>
<a name="SPLICE_DONOR_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>SPLICE_DONOR_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> SPLICE_DONOR_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001575">SO:0001575</a> A splice variant
 that changes the 2 base pair region at the 5' end of an intron (is a <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#SPLICE_REGION_VARIANT"><code>SPLICE_REGION_VARIANT</code></a>).</div>
</li>
</ul>
<a name="RARE_AMINO_ACID_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>RARE_AMINO_ACID_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> RARE_AMINO_ACID_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002008">SO:0002008</a> A sequence variant
 whereby at least one base of a codon encoding a rare amino acid is changed, resulting in a different encoded
 amino acid (children: selenocysteine_loss, pyrrolysine_loss).

 <b>Not</b> used in Jannovar annotations.</div>
</li>
</ul>
<a name="Z:Z_SMALLEST_HIGH_IMPACT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>_SMALLEST_HIGH_IMPACT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> _SMALLEST_HIGH_IMPACT</pre>
<div class="block">Marker for smallest <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation"><code>VariantEffect</code></a> with <a href="../../../../../de/charite/compbio/jannovar/annotation/PutativeImpact.html#HIGH"><code>PutativeImpact.HIGH</code></a> impact.</div>
</li>
</ul>
<a name="MISSENSE_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>MISSENSE_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> MISSENSE_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001583">SO:0001583</a> A sequence variant,
 that changes one or more bases, resulting in a different amino acid sequence but where the length is preserved.</div>
</li>
</ul>
<a name="INFRAME_INSERTION">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>INFRAME_INSERTION</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> INFRAME_INSERTION</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001821">SO:0001821</a> An inframe non
 synonymous variant that inserts bases into in the coding sequence (is a: inframe_indel,
 internal_feature_elongation).</div>
</li>
</ul>
<a name="DISRUPTIVE_INFRAME_INSERTION">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>DISRUPTIVE_INFRAME_INSERTION</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> DISRUPTIVE_INFRAME_INSERTION</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001824">SO:0001824</a> An inframe increase
 in cds length that inserts one or more codons into the coding sequence within an existing codon (is a:
 <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#INFRAME_INSERTION"><code>INFRAME_INSERTION</code></a>).</div>
</li>
</ul>
<a name="INFRAME_DELETION">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>INFRAME_DELETION</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> INFRAME_DELETION</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001822">SO:0001822</a> An inframe non
 synonymous variant that deletes bases from the coding sequence (is a: inframe_indel, feature_truncation).</div>
</li>
</ul>
<a name="DISRUPTIVE_INFRAME_DELETION">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>DISRUPTIVE_INFRAME_DELETION</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> DISRUPTIVE_INFRAME_DELETION</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001826">SO:0001826</a> An inframe decrease
 in cds length that deletes bases from the coding sequence starting within an existing codon (is a:
 <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#INFRAME_DELETION"><code>INFRAME_DELETION</code></a>).</div>
</li>
</ul>
<a name="FIVE_PRIME_UTR_TRUNCATION">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>FIVE_PRIME_UTR_TRUNCATION</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> FIVE_PRIME_UTR_TRUNCATION</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002013">SO:0002013</a> A sequence variant
 that causes the reduction of a the 5'UTR with regard to the reference sequence (is a:
 <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#FIVE_PRIME_UTR_EXON_VARIANT"><code>FIVE_PRIME_UTR_EXON_VARIANT</code></a> or <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#FIVE_PRIME_UTR_INTRON_VARIANT"><code>FIVE_PRIME_UTR_INTRON_VARIANT</code></a>)

 Jannovar does <b>not</b> yield use this at the moment.</div>
</li>
</ul>
<a name="THREE_PRIME_UTR_TRUNCATION">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>THREE_PRIME_UTR_TRUNCATION</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> THREE_PRIME_UTR_TRUNCATION</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002015">SO:0002015</a> A sequence variant
 that causes the reduction of a the 3' UTR with regard to the reference sequence (is a:
 <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#FIVE_PRIME_UTR_EXON_VARIANT"><code>FIVE_PRIME_UTR_EXON_VARIANT</code></a> or <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#FIVE_PRIME_UTR_INTRON_VARIANT"><code>FIVE_PRIME_UTR_INTRON_VARIANT</code></a>).

 Jannovar does <b>not</b> yield use this at the moment.</div>
</li>
</ul>
<a name="SPLICE_REGION_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>SPLICE_REGION_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> SPLICE_REGION_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001630">SO:0001630</a> A sequence variant
 in which a change has occurred within the region of the splice site, either within 1-3 bases of the exon or 3-8
 bases of the intron (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#SPLICING_VARIANT"><code>SPLICING_VARIANT</code></a>).</div>
</li>
</ul>
<a name="Z:Z_SMALLEST_MODERATE_IMPACT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>_SMALLEST_MODERATE_IMPACT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> _SMALLEST_MODERATE_IMPACT</pre>
<div class="block">Marker for smallest <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation"><code>VariantEffect</code></a> with <a href="../../../../../de/charite/compbio/jannovar/annotation/PutativeImpact.html#MODERATE"><code>PutativeImpact.MODERATE</code></a> impact.</div>
</li>
</ul>
<a name="STOP_RETAINED_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>STOP_RETAINED_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> STOP_RETAINED_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001567">SO:0001567</a> A sequence variant
 where at least one base in the terminator codon is changed, but the terminator remains (is a:
 <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#SYNONYMOUS_VARIANT"><code>SYNONYMOUS_VARIANT</code></a>, terminator_codon_variant).</div>
</li>
</ul>
<a name="INITIATOR_CODON_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>INITIATOR_CODON_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> INITIATOR_CODON_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001582">SO:0001582</a> A codon variant that
 changes at least one base of the first codon of a transcript (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#CODING_SEQUENCE_VARIANT"><code>CODING_SEQUENCE_VARIANT</code></a>, children:
 start_retained_variant, start_lost).</div>
</li>
</ul>
<a name="SYNONYMOUS_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>SYNONYMOUS_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> SYNONYMOUS_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001819">SO:0001819</a> A sequence variant
 where there is no resulting change to the encoded amino acid (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#CODING_SEQUENCE_VARIANT"><code>CODING_SEQUENCE_VARIANT</code></a>, children:
 start_retained_variant, stop_retained_variant).</div>
</li>
</ul>
<a name="CODING_TRANSCRIPT_INTRON_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>CODING_TRANSCRIPT_INTRON_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> CODING_TRANSCRIPT_INTRON_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001969">SO:0001969</a> A sequence variant
 that changes non-coding intro sequence in a non-coding transcript (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#CODING_TRANSCRIPT_VARIANT"><code>CODING_TRANSCRIPT_VARIANT</code></a>,
 <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#INTRON_VARIANT"><code>INTRON_VARIANT</code></a>).</div>
</li>
</ul>
<a name="NON_CODING_TRANSCRIPT_EXON_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>NON_CODING_TRANSCRIPT_EXON_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> NON_CODING_TRANSCRIPT_EXON_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001792">SO:0001792</a> A sequence variant
 that changes non-coding exon sequence in a non-coding transcript (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#NON_CODING_TRANSCRIPT_VARIANT"><code>NON_CODING_TRANSCRIPT_VARIANT</code></a>,
 <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#EXON_VARIANT"><code>EXON_VARIANT</code></a>).</div>
</li>
</ul>
<a name="NON_CODING_TRANSCRIPT_INTRON_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>NON_CODING_TRANSCRIPT_INTRON_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> NON_CODING_TRANSCRIPT_INTRON_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001970">SO:0001970</a> A sequence variant
 that changes non-coding intro sequence in a non-coding transcript (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#NON_CODING_TRANSCRIPT_VARIANT"><code>NON_CODING_TRANSCRIPT_VARIANT</code></a>,
 <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#INTRON_VARIANT"><code>INTRON_VARIANT</code></a>).</div>
</li>
</ul>
<a name="FIVE_PRIME_UTR_PREMATURE_START_CODON_GAIN_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>FIVE_PRIME_UTR_PREMATURE_START_CODON_GAIN_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> FIVE_PRIME_UTR_PREMATURE_START_CODON_GAIN_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001983">SO:0001983</a> A 5' UTR variant
 where a premature start codon is introduced, moved or lost (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#FIVE_PRIME_UTR_EXON_VARIANT"><code>FIVE_PRIME_UTR_EXON_VARIANT</code></a> or
 <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#FIVE_PRIME_UTR_INTRON_VARIANT"><code>FIVE_PRIME_UTR_INTRON_VARIANT</code></a>).

 <b>Not</b> used in Jannovar annotations.</div>
</li>
</ul>
<a name="FIVE_PRIME_UTR_EXON_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>FIVE_PRIME_UTR_EXON_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> FIVE_PRIME_UTR_EXON_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002092">SO:0002092</a> A UTR variant of the
 5' UTR (is a: 5_prime_UTR_variant; is a: UTR_variant).</div>
</li>
</ul>
<a name="THREE_PRIME_UTR_EXON_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>THREE_PRIME_UTR_EXON_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> THREE_PRIME_UTR_EXON_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002089">SO:0002089</a> A UTR variant of the
 3' UTR (is a: 3_prime_UTR_variant; is a: UTR_variant).</div>
</li>
</ul>
<a name="FIVE_PRIME_UTR_INTRON_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>FIVE_PRIME_UTR_INTRON_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> FIVE_PRIME_UTR_INTRON_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002091">SO:0002091</a> A UTR variant
 between 5' UTRs (is a: 5_prime_UTR_variant; is a: UTR_variant).</div>
</li>
</ul>
<a name="THREE_PRIME_UTR_INTRON_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>THREE_PRIME_UTR_INTRON_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> THREE_PRIME_UTR_INTRON_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002090">SO:0002090</a> A UTR variant
 between 3' UTRs (is a: 3_prime_UTR_variant; is a: UTR_variant).</div>
</li>
</ul>
<a name="Z:Z_SMALLEST_LOW_IMPACT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>_SMALLEST_LOW_IMPACT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> _SMALLEST_LOW_IMPACT</pre>
<div class="block">Marker for smallest <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation"><code>VariantEffect</code></a> with <a href="../../../../../de/charite/compbio/jannovar/annotation/PutativeImpact.html#LOW"><code>PutativeImpact.LOW</code></a> impact.</div>
</li>
</ul>
<a name="DIRECT_TANDEM_DUPLICATION">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>DIRECT_TANDEM_DUPLICATION</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> DIRECT_TANDEM_DUPLICATION</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:1000039">SO:1000039</a> A tandem duplication
 where the individual regions are in the same orientation (is a: tandem_duplication).

 In Jannovar used, as an additional marker to describe that an insertion is a duplication.</div>
</li>
</ul>
<a name="CUSTOM">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>CUSTOM</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> CUSTOM</pre>
<div class="block">Variant in a user-specified custom region.

 <b>Not</b> used in Jannovar annotations.</div>
</li>
</ul>
<a name="UPSTREAM_GENE_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>UPSTREAM_GENE_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> UPSTREAM_GENE_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001631">SO:0001631</a> A sequence variant
 located 5' of a gene (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#INTERGENIC_VARIANT"><code>INTERGENIC_VARIANT</code></a>).</div>
</li>
</ul>
<a name="DOWNSTREAM_GENE_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>DOWNSTREAM_GENE_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> DOWNSTREAM_GENE_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001632">SO:0001632</a> A sequence variant
 located 3' of a gene (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#INTERGENIC_VARIANT"><code>INTERGENIC_VARIANT</code></a>).</div>
</li>
</ul>
<a name="INTERGENIC_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>INTERGENIC_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> INTERGENIC_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001628">SO:0001628</a> A sequence variant
 located in the intergenic region, between genes (is a: feature_variant).</div>
</li>
</ul>
<a name="TF_BINDING_SITE_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>TF_BINDING_SITE_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> TF_BINDING_SITE_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001782">SO:0001782</a> A sequence variant
 located within a transcription factor binding site (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#REGULATORY_REGION_VARIANT"><code>REGULATORY_REGION_VARIANT</code></a>).

 <b>Not</b> used in Jannovar annotations.</div>
</li>
</ul>
<a name="REGULATORY_REGION_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>REGULATORY_REGION_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> REGULATORY_REGION_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001566">SO:0001566</a> A sequence variant
 located within a regulatory region (is a: feature_variant).

 <b>Not</b> used in Jannovar annotations.</div>
</li>
</ul>
<a name="CONSERVED_INTRON_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>CONSERVED_INTRON_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> CONSERVED_INTRON_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002018">SO:0002018</a> A transcript variant
 occurring within a conserved region of an intron (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#INTRON_VARIANT"><code>INTRON_VARIANT</code></a>).

 <b>Not</b> used in Jannovar annotations.</div>
</li>
</ul>
<a name="INTRAGENIC_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>INTRAGENIC_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> INTRAGENIC_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002011">SO:0002011</a> A variant that
 occurs within a gene but falls outside of all transcript features. This occurs when alternate transcripts of a
 gene do not share overlapping sequence (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#TRANSCRIPT_VARIANT"><code>TRANSCRIPT_VARIANT</code></a> ).</div>
</li>
</ul>
<a name="CONSERVED_INTERGENIC_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>CONSERVED_INTERGENIC_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> CONSERVED_INTERGENIC_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002017">SO:0002017</a> A sequence variant
 located in a conserved intergenic region, between genes (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#INTERGENIC_VARIANT"><code>INTERGENIC_VARIANT</code></a>).</div>
</li>
</ul>
<a name="STRUCTURAL_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>STRUCTURAL_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> STRUCTURAL_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001537">SO:0001537</a> A sequence variant
 that changes one or more sequence features (is a: sequence variant).</div>
</li>
</ul>
<a name="CODING_SEQUENCE_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>CODING_SEQUENCE_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> CODING_SEQUENCE_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001580">SO:0001580</a> A sequence variant
 that changes the coding sequence (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#CODING_TRANSCRIPT_VARIANT"><code>CODING_TRANSCRIPT_VARIANT</code></a>, <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#EXON_VARIANT"><code>EXON_VARIANT</code></a>).

 Sequence Ontology does <b>not</b> have a term
 <a href="http://www.sequenceontology.org/browser/current_svn/term/CODING_TRANSCRIPT_EXON_VARIANT" >
 CODING_TRANSCRIPT_EXON_VARIANT</a>, so we use this.

 <b>Not</b> used in Jannovar annotations.</div>
</li>
</ul>
<a name="INTRON_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>INTRON_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> INTRON_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001627">SO:0001627</a> A transcript variant
 occurring within an intron (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#TRANSCRIPT_VARIANT"><code>TRANSCRIPT_VARIANT</code></a>).

 Jannovar uses <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#CODING_TRANSCRIPT_INTRON_VARIANT"><code>CODING_TRANSCRIPT_INTRON_VARIANT</code></a> and <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#NON_CODING_TRANSCRIPT_INTRON_VARIANT"><code>NON_CODING_TRANSCRIPT_INTRON_VARIANT</code></a>
 instead.</div>
</li>
</ul>
<a name="EXON_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>EXON_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> EXON_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001791">SO:0001791</a> A sequence variant
 that changes exon sequence (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#TRANSCRIPT_VARIANT"><code>TRANSCRIPT_VARIANT</code></a>).</div>
</li>
</ul>
<a name="SPLICING_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>SPLICING_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> SPLICING_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001568">SO:0001568</a> A sequence variant
 that changes the process of splicing (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#GENE_VARIANT"><code>GENE_VARIANT</code></a>).

 <b>Not</b> used in Jannovar annotations.</div>
</li>
</ul>
<a name="MIRNA">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>MIRNA</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> MIRNA</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0000276">SO:0000276</a> Variant affects a
 miRNA (is a: miRNA_primary_transcript, small_regulatory_ncRNA).

 <b>Not</b> used in Jannovar annotations.</div>
</li>
</ul>
<a name="GENE_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>GENE_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> GENE_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001564">SO:0001564</a> A sequence variant
 where the structure of the gene is changed (is a: feature_variant).

 <b>Not</b> used in Jannovar annotations.</div>
</li>
</ul>
<a name="CODING_TRANSCRIPT_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>CODING_TRANSCRIPT_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> CODING_TRANSCRIPT_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001968">SO:0001968</a> A transcript variant
 of a protein coding gene (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#TRANSCRIPT_VARIANT"><code>TRANSCRIPT_VARIANT</code></a>).

 <b>Not</b> used in Jannovar annotations.</div>
</li>
</ul>
<a name="NON_CODING_TRANSCRIPT_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>NON_CODING_TRANSCRIPT_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> NON_CODING_TRANSCRIPT_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001619">SO:0001619</a> (is a:
 <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#TRANSCRIPT_VARIANT"><code>TRANSCRIPT_VARIANT</code></a>).

 Used for marking splicing variants as non-coding.</div>
</li>
</ul>
<a name="TRANSCRIPT_VARIANT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>TRANSCRIPT_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> TRANSCRIPT_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001576">SO:0001576</a> A sequence variant
 that changes the structure of the transcript (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#GENE_VARIANT"><code>GENE_VARIANT</code></a>). TRANSCRIPT_VARIANT, /**
 <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:">SO:</a> (is a: <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#GENE_VARIANT"><code>GENE_VARIANT</code></a>)).

 <b>Not</b> used in Jannovar annotations.</div>
</li>
</ul>
<a name="INTERGENIC_REGION">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>INTERGENIC_REGION</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> INTERGENIC_REGION</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0000605">SO:0000605</a> A region containing
 or overlapping no genes that is bounded on either side by a gene, or bounded by a gene and the end of the
 chromosome (is a: biological_region).

 <b>Not</b> used in Jannovar annotations.</div>
</li>
</ul>
<a name="CHROMOSOME">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>CHROMOSOME</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> CHROMOSOME</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0000340">SO:0000340</a> Structural unit
 composed of a nucleic acid molecule which controls its own replication through the interaction of specific
 proteins at one or more origins of replication (is a: replicon).

 <b>Not</b> used in Jannovar annotations.</div>
</li>
</ul>
<a name="SEQUENCE_VARIANT">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>SEQUENCE_VARIANT</h4>
<pre>public static final&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a> SEQUENCE_VARIANT</pre>
<div class="block"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001060">SO:0001060</a> Top level term for
 variants, can be used for marking "uknown effect".

 <b>Not</b> used in Jannovar annotations.</div>
</li>
</ul>
</li>
</ul>
<!-- ============ FIELD DETAIL =========== -->
<ul class="blockList">
<li class="blockList"><a name="field.detail">
<!--   -->
</a>
<h3>Field Detail</h3>
<a name="IS_SPLICING">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>IS_SPLICING</h4>
<pre>public static final&nbsp;com.google.common.base.Predicate&lt;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a>&gt; IS_SPLICING</pre>
<div class="block"><code>Predicate</code> for testing whether a <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation"><code>VariantEffect</code></a> is related to splicing.</div>
</li>
</ul>
<a name="IS_INTRONIC">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>IS_INTRONIC</h4>
<pre>public static final&nbsp;com.google.common.base.Predicate&lt;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a>&gt; IS_INTRONIC</pre>
<div class="block"><code>Predicate</code> for testing whether a <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation"><code>VariantEffect</code></a> is intronic.</div>
</li>
</ul>
<a name="TO_SO_TERM">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>TO_SO_TERM</h4>
<pre>public static final&nbsp;com.google.common.base.Function&lt;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a>,<a href="http://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&gt; TO_SO_TERM</pre>
<div class="block"><code>Function</code> for converting from <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation"><code>VariantEffect</code></a> to SO term String.</div>
</li>
</ul>
<a name="TO_LEGACY_NAME">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>TO_LEGACY_NAME</h4>
<pre>public static final&nbsp;com.google.common.base.Function&lt;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a>,<a href="http://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&gt; TO_LEGACY_NAME</pre>
<div class="block"><code>Function</code> for converting from <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation"><code>VariantEffect</code></a> to legacy term.</div>
</li>
</ul>
</li>
</ul>
<!-- ============ METHOD DETAIL ========== -->
<ul class="blockList">
<li class="blockList"><a name="method.detail">
<!--   -->
</a>
<h3>Method Detail</h3>
<a name="values--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>values</h4>
<pre>public static&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a>[]&nbsp;values()</pre>
<div class="block">Returns an array containing the constants of this enum type, in
the order they are declared.  This method may be used to iterate
over the constants as follows:
<pre>
for (VariantEffect c : VariantEffect.values())
&nbsp;   System.out.println(c);
</pre></div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>an array containing the constants of this enum type, in the order they are declared</dd>
</dl>
</li>
</ul>
<a name="valueOf-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>valueOf</h4>
<pre>public static&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation">VariantEffect</a>&nbsp;valueOf(<a href="http://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;name)</pre>
<div class="block">Returns the enum constant of this type with the specified name.
The string must match <i>exactly</i> an identifier used to declare an
enum constant in this type.  (Extraneous whitespace characters are 
not permitted.)</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>name</code> - the name of the enum constant to be returned.</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the enum constant with the specified name</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code><a href="http://docs.oracle.com/javase/8/docs/api/java/lang/IllegalArgumentException.html?is-external=true" title="class or interface in java.lang">IllegalArgumentException</a></code> - if this enum type has no constant with the specified name</dd>
<dd><code><a href="http://docs.oracle.com/javase/8/docs/api/java/lang/NullPointerException.html?is-external=true" title="class or interface in java.lang">NullPointerException</a></code> - if the argument is null</dd>
</dl>
</li>
</ul>
<a name="getLegacyTerm--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getLegacyTerm</h4>
<pre>public&nbsp;<a href="http://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;getLegacyTerm()</pre>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>Legacy (old Jannovar) id.</dd>
</dl>
</li>
</ul>
<a name="getImpact--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getImpact</h4>
<pre>public&nbsp;<a href="../../../../../de/charite/compbio/jannovar/annotation/PutativeImpact.html" title="enum in de.charite.compbio.jannovar.annotation">PutativeImpact</a>&nbsp;getImpact()</pre>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd><a href="../../../../../de/charite/compbio/jannovar/annotation/PutativeImpact.html" title="enum in de.charite.compbio.jannovar.annotation"><code>PutativeImpact</code></a> of this effect annotation.</dd>
</dl>
</li>
</ul>
<a name="getSequenceOntologyTerm--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getSequenceOntologyTerm</h4>
<pre>public&nbsp;<a href="http://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;getSequenceOntologyTerm()</pre>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd><a href="http://www.sequenceontology.org/browser/current_svn/term/String">String</a> with the Sequence
         Ontology term.</dd>
</dl>
</li>
</ul>
<a name="getSequenceOID--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getSequenceOID</h4>
<pre>public&nbsp;<a href="http://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;getSequenceOID()</pre>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd><a href="http://www.sequenceontology.org/browser/current_svn/term/String">String</a> with the Sequence
         Ontology ID.</dd>
</dl>
</li>
</ul>
<a name="getNumber--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getNumber</h4>
<pre>public&nbsp;int&nbsp;getNumber()</pre>
<div class="block">Forward to <code>ordinal()</code> member function.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd><code>int</code> with the number used for sorting values of type <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation"><code>VariantEffect</code></a>.</dd>
</dl>
</li>
</ul>
<a name="isStructural--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isStructural</h4>
<pre>public&nbsp;boolean&nbsp;isStructural()</pre>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd><code>true</code> if this <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation"><code>VariantEffect</code></a> annotates structural variants.</dd>
</dl>
</li>
</ul>
<a name="isSplicing--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isSplicing</h4>
<pre>public&nbsp;boolean&nbsp;isSplicing()</pre>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd><code>true</code> if this <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html" title="enum in de.charite.compbio.jannovar.annotation"><code>VariantEffect</code></a> could affect splicing.</dd>
</dl>
</li>
</ul>
<a name="isIntronic--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isIntronic</h4>
<pre>public&nbsp;boolean&nbsp;isIntronic()</pre>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd><code>true</code> if equal to <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#CODING_TRANSCRIPT_INTRON_VARIANT"><code>CODING_TRANSCRIPT_INTRON_VARIANT</code></a> or
         <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#NON_CODING_TRANSCRIPT_INTRON_VARIANT"><code>NON_CODING_TRANSCRIPT_INTRON_VARIANT</code></a>.</dd>
</dl>
</li>
</ul>
<a name="isOffExome--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isOffExome</h4>
<pre>public&nbsp;boolean&nbsp;isOffExome()</pre>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd><code>true</code> if the variant effect does not indicate a change affecting the exome, <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#CUSTOM"><code>CUSTOM</code></a>
         is considered on-exome</dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#isOffTranscript--"><code>isOffTranscript()</code></a></dd>
</dl>
</li>
</ul>
<a name="isOffTranscript--">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>isOffTranscript</h4>
<pre>public&nbsp;boolean&nbsp;isOffTranscript()</pre>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd><code>true</code> if the variant effect does not indicate a change affecting a transcript,
         <a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#CUSTOM"><code>CUSTOM</code></a> is considered on-transcript</dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantEffect.html#isOffExome--"><code>isOffExome()</code></a></dd>
</dl>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<!-- ========= END OF CLASS DATA ========= -->
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/VariantEffect.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../index-all.html">Index</a></li>
</ul>
<div class="aboutLanguage">Jannovar, 0.17</div>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../../../../../de/charite/compbio/jannovar/annotation/VariantAnnotator.html" title="class in de.charite.compbio.jannovar.annotation"><span class="typeNameLink">Prev&nbsp;Class</span></a></li>
<li>Next&nbsp;Class</li>
</ul>
<ul class="navList">
<li><a href="../../../../../index.html?de/charite/compbio/jannovar/annotation/VariantEffect.html" target="_top">Frames</a></li>
<li><a href="VariantEffect.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../../../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#enum.constant.summary">Enum Constants</a>&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li><a href="#enum.constant.detail">Enum Constants</a>&nbsp;|&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
<p class="legalCopy"><small>Copyright &#169; 2016. All rights reserved.</small></p>
</body>
</html>