<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (version 1.7.0_80) on Wed Dec 09 17:09:48 CET 2015 -->
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<title>Uses of Class de.charite.compbio.jannovar.reference.GenomeInterval (Jannovar 0.16-SNAPSHOT API)</title>
<meta name="date" content="2015-12-09">
<link rel="stylesheet" type="text/css" href="../../../../../../stylesheet.css" title="Style">
</head>
<body>
<script type="text/javascript"><!--
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="Uses of Class de.charite.compbio.jannovar.reference.GenomeInterval (Jannovar 0.16-SNAPSHOT API)";
    }
//-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar_top">
<!--   -->
</a><a href="#skip-navbar_top" title="Skip navigation links"></a><a name="navbar_top_firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../../overview-summary.html">Overview</a></li>
<li><a href="../package-summary.html">Package</a></li>
<li><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">Class</a></li>
<li class="navBarCell1Rev">Use</li>
<li><a href="../package-tree.html">Tree</a></li>
<li><a href="../../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../../index-all.html">Index</a></li>
<li><a href="../../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev</li>
<li>Next</li>
</ul>
<ul class="navList">
<li><a href="../../../../../../index.html?de/charite/compbio/jannovar/reference/class-use/GenomeInterval.html" target="_top">Frames</a></li>
<li><a href="GenomeInterval.html" target="_top">No Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../../../../../../allclasses-noframe.html">All Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip-navbar_top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<div class="header">
<h2 title="Uses of Class de.charite.compbio.jannovar.reference.GenomeInterval" class="title">Uses of Class<br>de.charite.compbio.jannovar.reference.GenomeInterval</h2>
</div>
<div class="classUseContainer">
<ul class="blockList">
<li class="blockList">
<table border="0" cellpadding="3" cellspacing="0" summary="Use table, listing packages, and an explanation">
<caption><span>Packages that use <a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Package</th>
<th class="colLast" scope="col">Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><a href="#de.charite.compbio.jannovar.hgvs.bridge">de.charite.compbio.jannovar.hgvs.bridge</a></td>
<td class="colLast">&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#de.charite.compbio.jannovar.htsjdk">de.charite.compbio.jannovar.htsjdk</a></td>
<td class="colLast">&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#de.charite.compbio.jannovar.reference">de.charite.compbio.jannovar.reference</a></td>
<td class="colLast">&nbsp;</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList">
<ul class="blockList">
<li class="blockList"><a name="de.charite.compbio.jannovar.hgvs.bridge">
<!--   -->
</a>
<h3>Uses of <a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a> in <a href="../../../../../../de/charite/compbio/jannovar/hgvs/bridge/package-summary.html">de.charite.compbio.jannovar.hgvs.bridge</a></h3>
<table border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../de/charite/compbio/jannovar/hgvs/bridge/package-summary.html">de.charite.compbio.jannovar.hgvs.bridge</a> that return <a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></code></td>
<td class="colLast"><span class="strong">NucleotideLocationConverter.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/hgvs/bridge/NucleotideLocationConverter.html#translateNucleotideRange(de.charite.compbio.jannovar.reference.TranscriptModel,%20de.charite.compbio.jannovar.hgvs.nts.NucleotideRange,%20de.charite.compbio.jannovar.hgvs.SequenceType)">translateNucleotideRange</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptModel.html" title="class in de.charite.compbio.jannovar.reference">TranscriptModel</a>&nbsp;tm,
                        <a href="../../../../../../de/charite/compbio/jannovar/hgvs/nts/NucleotideRange.html" title="class in de.charite.compbio.jannovar.hgvs.nts">NucleotideRange</a>&nbsp;range,
                        <a href="../../../../../../de/charite/compbio/jannovar/hgvs/SequenceType.html" title="enum in de.charite.compbio.jannovar.hgvs">SequenceType</a>&nbsp;sequenceType)</code>
<div class="block">Convert <a href="../../../../../../de/charite/compbio/jannovar/hgvs/nts/NucleotideRange.html" title="class in de.charite.compbio.jannovar.hgvs.nts"><code>NucleotideRange</code></a> on a <a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptModel.html" title="class in de.charite.compbio.jannovar.reference"><code>TranscriptModel</code></a> to a <a href="../../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomePosition</code></a></div>
</td>
</tr>
</tbody>
</table>
<table border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../de/charite/compbio/jannovar/hgvs/bridge/package-summary.html">de.charite.compbio.jannovar.hgvs.bridge</a> with parameters of type <a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>protected <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a></code></td>
<td class="colLast"><span class="strong">NucleotideChangeToGenomeVariantTranslationImplBase.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/hgvs/bridge/NucleotideChangeToGenomeVariantTranslationImplBase.html#getGenomeSeq(de.charite.compbio.jannovar.reference.Strand,%20de.charite.compbio.jannovar.reference.GenomeInterval)">getGenomeSeq</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/Strand.html" title="enum in de.charite.compbio.jannovar.reference">Strand</a>&nbsp;strand,
            <a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;gItv)</code>
<div class="block">Return sequence from the reference in the given interval <code>gItv</code>.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="de.charite.compbio.jannovar.htsjdk">
<!--   -->
</a>
<h3>Uses of <a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a> in <a href="../../../../../../de/charite/compbio/jannovar/htsjdk/package-summary.html">de.charite.compbio.jannovar.htsjdk</a></h3>
<table border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../de/charite/compbio/jannovar/htsjdk/package-summary.html">de.charite.compbio.jannovar.htsjdk</a> with parameters of type <a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code><a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a></code></td>
<td class="colLast"><span class="strong">GenomeRegionSequenceExtractor.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/htsjdk/GenomeRegionSequenceExtractor.html#load(de.charite.compbio.jannovar.reference.GenomeInterval)">load</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;region)</code>
<div class="block">Load sequence from the given <code>region</code> from <a href="../../../../../../de/charite/compbio/jannovar/htsjdk/GenomeRegionSequenceExtractor.html#indexedFile"><code>GenomeRegionSequenceExtractor.indexedFile</code></a></div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="de.charite.compbio.jannovar.reference">
<!--   -->
</a>
<h3>Uses of <a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a> in <a href="../../../../../../de/charite/compbio/jannovar/reference/package-summary.html">de.charite.compbio.jannovar.reference</a></h3>
<table border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../de/charite/compbio/jannovar/reference/package-summary.html">de.charite.compbio.jannovar.reference</a> that return <a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></code></td>
<td class="colLast"><span class="strong">TranscriptModelBuilder.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptModelBuilder.html#getCDSRegion()">getCDSRegion</a></strong>()</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></code></td>
<td class="colLast"><span class="strong">TranscriptModel.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptModel.html#getCDSRegion()">getCDSRegion</a></strong>()</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></code></td>
<td class="colLast"><span class="strong">TranscriptSequenceOntologyDecorator.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#getDownstreamInterval()">getDownstreamInterval</a></strong>()</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></code></td>
<td class="colLast"><span class="strong">TranscriptSequenceOntologyDecorator.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#getFivePrimeUTRInterval()">getFivePrimeUTRInterval</a></strong>()</code>
<div class="block">Returns the <b>genomic</b> 5' UTR interval.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></code></td>
<td class="colLast"><span class="strong">GenomeVariant.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeVariant.html#getGenomeInterval()">getGenomeInterval</a></strong>()</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></code></td>
<td class="colLast"><span class="strong">TranscriptSequenceOntologyDecorator.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#getStartCodonInterval()">getStartCodonInterval</a></strong>()</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></code></td>
<td class="colLast"><span class="strong">TranscriptSequenceOntologyDecorator.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#getStopCodonInterval()">getStopCodonInterval</a></strong>()</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></code></td>
<td class="colLast"><span class="strong">TranscriptSequenceOntologyDecorator.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#getThreePrimeUTRInterval()">getThreePrimeUTRInterval</a></strong>()</code>
<div class="block">Returns the <b>genomic</b> 3' UTR interval.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></code></td>
<td class="colLast"><span class="strong">TranscriptModelBuilder.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptModelBuilder.html#getTXRegion()">getTXRegion</a></strong>()</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></code></td>
<td class="colLast"><span class="strong">TranscriptModel.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptModel.html#getTXRegion()">getTXRegion</a></strong>()</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></code></td>
<td class="colLast"><span class="strong">TranscriptSequenceOntologyDecorator.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#getUpstreamInterval()">getUpstreamInterval</a></strong>()</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></code></td>
<td class="colLast"><span class="strong">GenomeInterval.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html#intersection(de.charite.compbio.jannovar.reference.GenomeInterval)">intersection</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;other)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></code></td>
<td class="colLast"><span class="strong">TranscriptModel.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptModel.html#intronRegion(int)">intronRegion</a></strong>(int&nbsp;i)</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></code></td>
<td class="colLast"><span class="strong">GenomeInterval.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html#withMorePadding(int)">withMorePadding</a></strong>(int&nbsp;padding)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></code></td>
<td class="colLast"><span class="strong">GenomeInterval.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html#withMorePadding(int,%20int)">withMorePadding</a></strong>(int&nbsp;paddingUpstream,
               int&nbsp;paddingDownstream)</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></code></td>
<td class="colLast"><span class="strong">GenomeInterval.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html#withStrand(de.charite.compbio.jannovar.reference.Strand)">withStrand</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/Strand.html" title="enum in de.charite.compbio.jannovar.reference">Strand</a>&nbsp;strand)</code>
<div class="block">convert into GenomeInterval of the given strand</div>
</td>
</tr>
</tbody>
</table>
<table border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../de/charite/compbio/jannovar/reference/package-summary.html">de.charite.compbio.jannovar.reference</a> that return types with arguments of type <a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code><a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&gt;</code></td>
<td class="colLast"><span class="strong">TranscriptModelBuilder.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptModelBuilder.html#getExonRegions()">getExonRegions</a></strong>()</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>com.google.common.collect.ImmutableList&lt;<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&gt;</code></td>
<td class="colLast"><span class="strong">TranscriptModel.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptModel.html#getExonRegions()">getExonRegions</a></strong>()</code>&nbsp;</td>
</tr>
</tbody>
</table>
<table border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../de/charite/compbio/jannovar/reference/package-summary.html">de.charite.compbio.jannovar.reference</a> with parameters of type <a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><span class="strong">TranscriptModelBuilder.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptModelBuilder.html#addExonRegion(de.charite.compbio.jannovar.reference.GenomeInterval)">addExonRegion</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;exonRegion)</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="strong">GenomeInterval.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html#contains(de.charite.compbio.jannovar.reference.GenomeInterval)">contains</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;other)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="strong">TranscriptSequenceOntologyDecorator.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#containsExon(de.charite.compbio.jannovar.reference.GenomeInterval)">containsExon</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><span class="strong">GenomePosition.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html#differenceTo(de.charite.compbio.jannovar.reference.GenomeInterval)">differenceTo</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;itv)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></code></td>
<td class="colLast"><span class="strong">GenomeInterval.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html#intersection(de.charite.compbio.jannovar.reference.GenomeInterval)">intersection</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;other)</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="strong">TranscriptSequenceOntologyDecorator.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#liesInCDSExon(de.charite.compbio.jannovar.reference.GenomeInterval)">liesInCDSExon</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="strong">TranscriptSequenceOntologyDecorator.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#liesInExon(de.charite.compbio.jannovar.reference.GenomeInterval)">liesInExon</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="strong">TranscriptSequenceOntologyDecorator.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#liesInIntron(de.charite.compbio.jannovar.reference.GenomeInterval)">liesInIntron</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="strong">GenomeInterval.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html#overlapsWith(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWith</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;other)</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="strong">TranscriptSequenceOntologyDecorator.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#overlapsWithCDS(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWithCDS</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="strong">TranscriptSequenceOntologyDecorator.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#overlapsWithCDSExon(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWithCDSExon</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="strong">TranscriptSequenceOntologyDecorator.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#overlapsWithCDSIntron(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWithCDSIntron</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;changeInterval)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="strong">TranscriptSequenceOntologyDecorator.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#overlapsWithDownstreamRegion(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWithDownstreamRegion</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>
<div class="block">Returns whether the given <code>interval</code> overlaps with the downstream region of the transcript.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="strong">TranscriptSequenceOntologyDecorator.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#overlapsWithExon(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWithExon</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="strong">TranscriptSequenceOntologyDecorator.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#overlapsWithFivePrimeUTR(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWithFivePrimeUTR</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="strong">TranscriptSequenceOntologyDecorator.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#overlapsWithIntron(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWithIntron</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;changeInterval)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="strong">TranscriptSequenceOntologyDecorator.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#overlapsWithSpliceAcceptorSite(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWithSpliceAcceptorSite</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>
<div class="block">Returns whether the given <code>interval</code> overlaps with a splice acceptor site.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="strong">TranscriptSequenceOntologyDecorator.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#overlapsWithSpliceDonorSite(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWithSpliceDonorSite</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>
<div class="block">Returns whether the given <code>interval</code> overlaps with a splice donor site.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="strong">TranscriptSequenceOntologyDecorator.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#overlapsWithSpliceRegion(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWithSpliceRegion</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>
<div class="block">Returns whether the given <code>interval</code> overlaps with a splice region.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="strong">TranscriptSequenceOntologyDecorator.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#overlapsWithThreePrimeUTR(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWithThreePrimeUTR</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="strong">TranscriptSequenceOntologyDecorator.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#overlapsWithTranslationalStartSite(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWithTranslationalStartSite</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="strong">TranscriptSequenceOntologyDecorator.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#overlapsWithTranslationalStopSite(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWithTranslationalStopSite</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="strong">TranscriptSequenceOntologyDecorator.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#overlapsWithUpstreamRegion(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWithUpstreamRegion</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>
<div class="block">Returns whether the given <code>interval</code> overlaps with the upstream region of the transcript.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code><a href="../../../../../../de/charite/compbio/jannovar/reference/CDSInterval.html" title="class in de.charite.compbio.jannovar.reference">CDSInterval</a></code></td>
<td class="colLast"><span class="strong">TranscriptProjectionDecorator.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptProjectionDecorator.html#projectGenomeToCDSInterval(de.charite.compbio.jannovar.reference.GenomeInterval)">projectGenomeToCDSInterval</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>
<div class="block">Translate <a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> to <a href="../../../../../../de/charite/compbio/jannovar/reference/CDSInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>CDSInterval</code></a> for <a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptProjectionDecorator.html#transcript"><code>TranscriptProjectionDecorator.transcript</code></a>.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptInterval.html" title="class in de.charite.compbio.jannovar.reference">TranscriptInterval</a></code></td>
<td class="colLast"><span class="strong">TranscriptProjectionDecorator.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptProjectionDecorator.html#projectGenomeToTXInterval(de.charite.compbio.jannovar.reference.GenomeInterval)">projectGenomeToTXInterval</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>
<div class="block">Translate <a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> to <a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>TranscriptInterval</code></a> for <a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptProjectionDecorator.html#transcript"><code>TranscriptProjectionDecorator.transcript</code></a>.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><span class="strong">TranscriptModelBuilder.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptModelBuilder.html#setCDSRegion(de.charite.compbio.jannovar.reference.GenomeInterval)">setCDSRegion</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;cdsRegion)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><span class="strong">TranscriptModelBuilder.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptModelBuilder.html#setTXRegion(de.charite.compbio.jannovar.reference.GenomeInterval)">setTXRegion</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;txRegion)</code>&nbsp;</td>
</tr>
</tbody>
</table>
<table border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Method parameters in <a href="../../../../../../de/charite/compbio/jannovar/reference/package-summary.html">de.charite.compbio.jannovar.reference</a> with type arguments of type <a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><span class="strong">TranscriptModelBuilder.</span><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptModelBuilder.html#clearExonRegions(java.util.ArrayList)">clearExonRegions</a></strong>(<a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&gt;&nbsp;exonRegions)</code>&nbsp;</td>
</tr>
</tbody>
</table>
<table border="0" cellpadding="3" cellspacing="0" summary="Use table, listing constructors, and an explanation">
<caption><span>Constructors in <a href="../../../../../../de/charite/compbio/jannovar/reference/package-summary.html">de.charite.compbio.jannovar.reference</a> with parameters of type <a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colLast"><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html#GenomeInterval(de.charite.compbio.jannovar.reference.GenomeInterval)">GenomeInterval</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;other)</code>
<div class="block">construct genome interval from other with selected coordinate system</div>
</td>
</tr>
<tr class="rowColor">
<td class="colLast"><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html#GenomeInterval(de.charite.compbio.jannovar.reference.GenomeInterval,%20de.charite.compbio.jannovar.reference.Strand)">GenomeInterval</a></strong>(<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;other,
              <a href="../../../../../../de/charite/compbio/jannovar/reference/Strand.html" title="enum in de.charite.compbio.jannovar.reference">Strand</a>&nbsp;strand)</code>
<div class="block">construct genome interval from other with selected strand</div>
</td>
</tr>
<tr class="altColor">
<td class="colLast"><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptModel.html#TranscriptModel(java.lang.String,%20java.lang.String,%20de.charite.compbio.jannovar.reference.GenomeInterval,%20de.charite.compbio.jannovar.reference.GenomeInterval,%20com.google.common.collect.ImmutableList,%20java.lang.String,%20java.lang.String,%20int)">TranscriptModel</a></strong>(<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;accession,
               <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;geneSymbol,
               <a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;txRegion,
               <a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;cdsRegion,
               com.google.common.collect.ImmutableList&lt;<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&gt;&nbsp;exonRegions,
               <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;sequence,
               <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;geneID,
               int&nbsp;transcriptSupportLevel)</code>
<div class="block">Initialize the TranscriptInfo object from the given parameters.</div>
</td>
</tr>
</tbody>
</table>
<table border="0" cellpadding="3" cellspacing="0" summary="Use table, listing constructors, and an explanation">
<caption><span>Constructor parameters in <a href="../../../../../../de/charite/compbio/jannovar/reference/package-summary.html">de.charite.compbio.jannovar.reference</a> with type arguments of type <a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colLast"><code><strong><a href="../../../../../../de/charite/compbio/jannovar/reference/TranscriptModel.html#TranscriptModel(java.lang.String,%20java.lang.String,%20de.charite.compbio.jannovar.reference.GenomeInterval,%20de.charite.compbio.jannovar.reference.GenomeInterval,%20com.google.common.collect.ImmutableList,%20java.lang.String,%20java.lang.String,%20int)">TranscriptModel</a></strong>(<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;accession,
               <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;geneSymbol,
               <a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;txRegion,
               <a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;cdsRegion,
               com.google.common.collect.ImmutableList&lt;<a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&gt;&nbsp;exonRegions,
               <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;sequence,
               <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;geneID,
               int&nbsp;transcriptSupportLevel)</code>
<div class="block">Initialize the TranscriptInfo object from the given parameters.</div>
</td>
</tr>
</tbody>
</table>
</li>
</ul>
</li>
</ul>
</div>
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar_bottom">
<!--   -->
</a><a href="#skip-navbar_bottom" title="Skip navigation links"></a><a name="navbar_bottom_firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../../overview-summary.html">Overview</a></li>
<li><a href="../package-summary.html">Package</a></li>
<li><a href="../../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">Class</a></li>
<li class="navBarCell1Rev">Use</li>
<li><a href="../package-tree.html">Tree</a></li>
<li><a href="../../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../../index-all.html">Index</a></li>
<li><a href="../../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev</li>
<li>Next</li>
</ul>
<ul class="navList">
<li><a href="../../../../../../index.html?de/charite/compbio/jannovar/reference/class-use/GenomeInterval.html" target="_top">Frames</a></li>
<li><a href="GenomeInterval.html" target="_top">No Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../../../../../../allclasses-noframe.html">All Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip-navbar_bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
<p class="legalCopy"><small>Copyright &#169; 2015. All rights reserved.</small></p>
</body>
</html>
