<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (version 1.7.0_80) on Wed Dec 09 17:09:46 CET 2015 -->
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<title>TranscriptSequenceOntologyDecorator (Jannovar 0.16-SNAPSHOT API)</title>
<meta name="date" content="2015-12-09">
<link rel="stylesheet" type="text/css" href="../../../../../stylesheet.css" title="Style">
</head>
<body>
<script type="text/javascript"><!--
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="TranscriptSequenceOntologyDecorator (Jannovar 0.16-SNAPSHOT API)";
    }
//-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar_top">
<!--   -->
</a><a href="#skip-navbar_top" title="Skip navigation links"></a><a name="navbar_top_firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/TranscriptSequenceOntologyDecorator.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../index-all.html">Index</a></li>
<li><a href="../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceDecorator.html" title="class in de.charite.compbio.jannovar.reference"><span class="strong">Prev Class</span></a></li>
<li><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSupportLevels.html" title="interface in de.charite.compbio.jannovar.reference"><span class="strong">Next Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../../../../../index.html?de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html" target="_top">Frames</a></li>
<li><a href="TranscriptSequenceOntologyDecorator.html" target="_top">No Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../../../../../allclasses-noframe.html">All Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor_summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor_detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_detail">Method</a></li>
</ul>
</div>
<a name="skip-navbar_top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="subTitle">de.charite.compbio.jannovar.reference</div>
<h2 title="Class TranscriptSequenceOntologyDecorator" class="title">Class TranscriptSequenceOntologyDecorator</h2>
</div>
<div class="contentContainer">
<ul class="inheritance">
<li><a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Object.html?is-external=true" title="class or interface in java.lang">java.lang.Object</a></li>
<li>
<ul class="inheritance">
<li>de.charite.compbio.jannovar.reference.TranscriptSequenceOntologyDecorator</li>
</ul>
</li>
</ul>
<div class="description">
<ul class="blockList">
<li class="blockList">
<hr>
<br>
<pre>public final class <span class="strong">TranscriptSequenceOntologyDecorator</span>
extends <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Object.html?is-external=true" title="class or interface in java.lang">Object</a></pre>
<div class="block">Functionality for finding out about certain points/regions of <a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptModel.html" title="class in de.charite.compbio.jannovar.reference"><code>TranscriptModel</code></a> using <b>genomic</b> positions.</div>
<dl><dt><span class="strong">Author:</span></dt>
  <dd><a href="mailto:manuel.holtgrewe@charite.de">Manuel Holtgrewe</a></dd></dl>
</li>
</ul>
</div>
<div class="summary">
<ul class="blockList">
<li class="blockList">
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor_summary">
<!--   -->
</a>
<h3>Constructor Summary</h3>
<table class="overviewSummary" border="0" cellpadding="3" cellspacing="0" summary="Constructor Summary table, listing constructors, and an explanation">
<caption><span>Constructors</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tr class="altColor">
<td class="colOne"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#TranscriptSequenceOntologyDecorator(de.charite.compbio.jannovar.reference.TranscriptModel)">TranscriptSequenceOntologyDecorator</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptModel.html" title="class in de.charite.compbio.jannovar.reference">TranscriptModel</a>&nbsp;transcript)</code>
<div class="block">Initialize the object with the given <a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptModel.html" title="class in de.charite.compbio.jannovar.reference"><code>TranscriptModel</code></a>.</div>
</td>
</tr>
</table>
</li>
</ul>
<!-- ========== METHOD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="method_summary">
<!--   -->
</a>
<h3>Method Summary</h3>
<table class="overviewSummary" border="0" cellpadding="3" cellspacing="0" summary="Method Summary table, listing methods, and an explanation">
<caption><span>Methods</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#containsExon(de.charite.compbio.jannovar.reference.GenomeInterval)">containsExon</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code><a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#getDownstreamInterval()">getDownstreamInterval</a></strong>()</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code><a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#getFivePrimeUTRInterval()">getFivePrimeUTRInterval</a></strong>()</code>
<div class="block">Returns the <b>genomic</b> 5' UTR interval.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code><a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#getStartCodonInterval()">getStartCodonInterval</a></strong>()</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code><a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#getStopCodonInterval()">getStopCodonInterval</a></strong>()</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code><a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#getThreePrimeUTRInterval()">getThreePrimeUTRInterval</a></strong>()</code>
<div class="block">Returns the <b>genomic</b> 3' UTR interval.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptModel.html" title="class in de.charite.compbio.jannovar.reference">TranscriptModel</a></code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#getTranscript()">getTranscript</a></strong>()</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code><a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a></code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#getUpstreamInterval()">getUpstreamInterval</a></strong>()</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#liesInCDS(de.charite.compbio.jannovar.reference.GenomePosition)">liesInCDS</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference">GenomePosition</a>&nbsp;pos)</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#liesInCDSExon(de.charite.compbio.jannovar.reference.GenomeInterval)">liesInCDSExon</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#liesInCDSExon(de.charite.compbio.jannovar.reference.GenomePosition)">liesInCDSExon</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference">GenomePosition</a>&nbsp;pos)</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#liesInCDSIntron(de.charite.compbio.jannovar.reference.GenomePosition)">liesInCDSIntron</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference">GenomePosition</a>&nbsp;pos)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#liesInDownstreamRegion(de.charite.compbio.jannovar.reference.GenomePosition)">liesInDownstreamRegion</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference">GenomePosition</a>&nbsp;pos)</code>
<div class="block">Returns whether the given <code>pos</code> lies within with the downstream region of the transcript.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#liesInExon(de.charite.compbio.jannovar.reference.GenomeInterval)">liesInExon</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#liesInExon(de.charite.compbio.jannovar.reference.GenomePosition)">liesInExon</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference">GenomePosition</a>&nbsp;pos)</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#liesInFivePrimeUTR(de.charite.compbio.jannovar.reference.GenomePosition)">liesInFivePrimeUTR</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference">GenomePosition</a>&nbsp;pos)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#liesInIntron(de.charite.compbio.jannovar.reference.GenomeInterval)">liesInIntron</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#liesInIntron(de.charite.compbio.jannovar.reference.GenomePosition)">liesInIntron</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference">GenomePosition</a>&nbsp;pos)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#liesInSpliceAcceptorSite(de.charite.compbio.jannovar.reference.GenomePosition)">liesInSpliceAcceptorSite</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference">GenomePosition</a>&nbsp;pos)</code>
<div class="block">Returns whether the given <code>pos</code> lies within a splice acceptor site.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#liesInSpliceDonorSite(de.charite.compbio.jannovar.reference.GenomePosition)">liesInSpliceDonorSite</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference">GenomePosition</a>&nbsp;pos)</code>
<div class="block">Returns whether the given <code>pos</code> lies within a splice donor site.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#liesInSpliceRegion(de.charite.compbio.jannovar.reference.GenomePosition)">liesInSpliceRegion</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference">GenomePosition</a>&nbsp;pos)</code>
<div class="block">Returns whether the given <code>pos</code> lies within a splice region.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#liesInThreePrimeUTR(de.charite.compbio.jannovar.reference.GenomePosition)">liesInThreePrimeUTR</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference">GenomePosition</a>&nbsp;pos)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#liesInTranslationalStartSite(de.charite.compbio.jannovar.reference.GenomePosition)">liesInTranslationalStartSite</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference">GenomePosition</a>&nbsp;pos)</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#liesInTranslationalStopSite(de.charite.compbio.jannovar.reference.GenomePosition)">liesInTranslationalStopSite</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference">GenomePosition</a>&nbsp;pos)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#liesInUpstreamRegion(de.charite.compbio.jannovar.reference.GenomePosition)">liesInUpstreamRegion</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference">GenomePosition</a>&nbsp;pos)</code>
<div class="block">Returns whether the given <code>pos</code> lies within with the upstream region of the transcript.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#overlapsWithCDS(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWithCDS</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#overlapsWithCDSExon(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWithCDSExon</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#overlapsWithCDSIntron(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWithCDSIntron</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;changeInterval)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#overlapsWithDownstreamRegion(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWithDownstreamRegion</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>
<div class="block">Returns whether the given <code>interval</code> overlaps with the downstream region of the transcript.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#overlapsWithExon(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWithExon</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#overlapsWithFivePrimeUTR(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWithFivePrimeUTR</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#overlapsWithIntron(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWithIntron</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;changeInterval)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#overlapsWithSpliceAcceptorSite(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWithSpliceAcceptorSite</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>
<div class="block">Returns whether the given <code>interval</code> overlaps with a splice acceptor site.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#overlapsWithSpliceDonorSite(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWithSpliceDonorSite</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>
<div class="block">Returns whether the given <code>interval</code> overlaps with a splice donor site.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#overlapsWithSpliceRegion(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWithSpliceRegion</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>
<div class="block">Returns whether the given <code>interval</code> overlaps with a splice region.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#overlapsWithThreePrimeUTR(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWithThreePrimeUTR</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#overlapsWithTranslationalStartSite(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWithTranslationalStartSite</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#overlapsWithTranslationalStopSite(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWithTranslationalStopSite</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#overlapsWithUpstreamRegion(de.charite.compbio.jannovar.reference.GenomeInterval)">overlapsWithUpstreamRegion</a></strong>(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</code>
<div class="block">Returns whether the given <code>interval</code> overlaps with the upstream region of the transcript.</div>
</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="methods_inherited_from_class_java.lang.Object">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;java.lang.<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Object.html?is-external=true" title="class or interface in java.lang">Object</a></h3>
<code><a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Object.html?is-external=true#clone()" title="class or interface in java.lang">clone</a>, <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Object.html?is-external=true#equals(java.lang.Object)" title="class or interface in java.lang">equals</a>, <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Object.html?is-external=true#finalize()" title="class or interface in java.lang">finalize</a>, <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Object.html?is-external=true#getClass()" title="class or interface in java.lang">getClass</a>, <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Object.html?is-external=true#hashCode()" title="class or interface in java.lang">hashCode</a>, <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Object.html?is-external=true#notify()" title="class or interface in java.lang">notify</a>, <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Object.html?is-external=true#notifyAll()" title="class or interface in java.lang">notifyAll</a>, <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Object.html?is-external=true#toString()" title="class or interface in java.lang">toString</a>, <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Object.html?is-external=true#wait()" title="class or interface in java.lang">wait</a>, <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Object.html?is-external=true#wait(long)" title="class or interface in java.lang">wait</a>, <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Object.html?is-external=true#wait(long,%20int)" title="class or interface in java.lang">wait</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="details">
<ul class="blockList">
<li class="blockList">
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor_detail">
<!--   -->
</a>
<h3>Constructor Detail</h3>
<a name="TranscriptSequenceOntologyDecorator(de.charite.compbio.jannovar.reference.TranscriptModel)">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>TranscriptSequenceOntologyDecorator</h4>
<pre>public&nbsp;TranscriptSequenceOntologyDecorator(<a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptModel.html" title="class in de.charite.compbio.jannovar.reference">TranscriptModel</a>&nbsp;transcript)</pre>
<div class="block">Initialize the object with the given <a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptModel.html" title="class in de.charite.compbio.jannovar.reference"><code>TranscriptModel</code></a>.</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>transcript</code> - the <a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptModel.html" title="class in de.charite.compbio.jannovar.reference"><code>TranscriptModel</code></a> to decorate</dd></dl>
</li>
</ul>
</li>
</ul>
<!-- ============ METHOD DETAIL ========== -->
<ul class="blockList">
<li class="blockList"><a name="method_detail">
<!--   -->
</a>
<h3>Method Detail</h3>
<a name="getTranscript()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getTranscript</h4>
<pre>public&nbsp;<a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptModel.html" title="class in de.charite.compbio.jannovar.reference">TranscriptModel</a>&nbsp;getTranscript()</pre>
<dl><dt><span class="strong">Returns:</span></dt><dd>the transcript</dd></dl>
</li>
</ul>
<a name="getStartCodonInterval()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getStartCodonInterval</h4>
<pre>public&nbsp;<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;getStartCodonInterval()</pre>
<dl><dt><span class="strong">Returns:</span></dt><dd>the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> with the start codon</dd></dl>
</li>
</ul>
<a name="getStopCodonInterval()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getStopCodonInterval</h4>
<pre>public&nbsp;<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;getStopCodonInterval()</pre>
<dl><dt><span class="strong">Returns:</span></dt><dd>the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> with the start codon</dd></dl>
</li>
</ul>
<a name="getFivePrimeUTRInterval()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getFivePrimeUTRInterval</h4>
<pre>public&nbsp;<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;getFivePrimeUTRInterval()</pre>
<div class="block">Returns the <b>genomic</b> 5' UTR interval.</div>
<dl><dt><span class="strong">Returns:</span></dt><dd>the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> with the 5' UTR</dd></dl>
</li>
</ul>
<a name="getThreePrimeUTRInterval()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getThreePrimeUTRInterval</h4>
<pre>public&nbsp;<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;getThreePrimeUTRInterval()</pre>
<div class="block">Returns the <b>genomic</b> 3' UTR interval.</div>
<dl><dt><span class="strong">Returns:</span></dt><dd>the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> with the 3' UTR</dd></dl>
</li>
</ul>
<a name="containsExon(de.charite.compbio.jannovar.reference.GenomeInterval)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>containsExon</h4>
<pre>public&nbsp;boolean&nbsp;containsExon(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</pre>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>interval</code> - query whether <code>interval</code> contains an exon</dd>
<dt><span class="strong">Returns:</span></dt><dd><code>true</code> if <code>interval</code> contains a full exon (coding or non-coding).</dd></dl>
</li>
</ul>
<a name="overlapsWithCDSExon(de.charite.compbio.jannovar.reference.GenomeInterval)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>overlapsWithCDSExon</h4>
<pre>public&nbsp;boolean&nbsp;overlapsWithCDSExon(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</pre>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>interval</code> - query whether <code>interval</code> overlaps with a CDS exon (exon that overlaps with CDS)</dd>
<dt><span class="strong">Returns:</span></dt><dd><code>true</code> if <code>interval</code> overlaps with a CDS-overlapping exon</dd></dl>
</li>
</ul>
<a name="liesInCDSExon(de.charite.compbio.jannovar.reference.GenomeInterval)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>liesInCDSExon</h4>
<pre>public&nbsp;boolean&nbsp;liesInCDSExon(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</pre>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>interval</code> - the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> to use for querying</dd>
<dt><span class="strong">Returns:</span></dt><dd><code>true</code> if the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> falls fully into the coding part of an exon</dd></dl>
</li>
</ul>
<a name="liesInCDSExon(de.charite.compbio.jannovar.reference.GenomePosition)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>liesInCDSExon</h4>
<pre>public&nbsp;boolean&nbsp;liesInCDSExon(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference">GenomePosition</a>&nbsp;pos)</pre>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>pos</code> - the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomePosition</code></a> to use for querying</dd>
<dt><span class="strong">Returns:</span></dt><dd><code>true</code> if the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomePosition</code></a> points to a base in the coding part of an exon</dd></dl>
</li>
</ul>
<a name="overlapsWithCDS(de.charite.compbio.jannovar.reference.GenomeInterval)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>overlapsWithCDS</h4>
<pre>public&nbsp;boolean&nbsp;overlapsWithCDS(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</pre>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>interval</code> - query whether <code>interval</code> overlaps with the CDS region</dd>
<dt><span class="strong">Returns:</span></dt><dd><code>true</code> if <code>interval</code> overlaps with the CDS region of the transcript</dd></dl>
</li>
</ul>
<a name="liesInCDS(de.charite.compbio.jannovar.reference.GenomePosition)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>liesInCDS</h4>
<pre>public&nbsp;boolean&nbsp;liesInCDS(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference">GenomePosition</a>&nbsp;pos)</pre>
<dl><dt><span class="strong">Returns:</span></dt><dd><code>true</code> if <code>pos</code> lies within the CDS of <a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#transcript"><code>transcript</code></a></dd></dl>
</li>
</ul>
<a name="overlapsWithIntron(de.charite.compbio.jannovar.reference.GenomeInterval)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>overlapsWithIntron</h4>
<pre>public&nbsp;boolean&nbsp;overlapsWithIntron(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;changeInterval)</pre>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>changeInterval</code> - the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> to use for the query</dd>
<dt><span class="strong">Returns:</span></dt><dd><code>true</code> if <code>changeInterval</code> overlaps with an intron of <a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#transcript"><code>transcript</code></a></dd></dl>
</li>
</ul>
<a name="liesInIntron(de.charite.compbio.jannovar.reference.GenomePosition)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>liesInIntron</h4>
<pre>public&nbsp;boolean&nbsp;liesInIntron(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference">GenomePosition</a>&nbsp;pos)</pre>
<dl><dt><span class="strong">Returns:</span></dt><dd><code>true</code> if <code>pos</code> lies within an intron of <a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#transcript"><code>transcript</code></a></dd></dl>
</li>
</ul>
<a name="overlapsWithCDSIntron(de.charite.compbio.jannovar.reference.GenomeInterval)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>overlapsWithCDSIntron</h4>
<pre>public&nbsp;boolean&nbsp;overlapsWithCDSIntron(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;changeInterval)</pre>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>changeInterval</code> - the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> to use for the query</dd>
<dt><span class="strong">Returns:</span></dt><dd><code>true</code> if <code>changeInterval</code> overlaps with an intron of <a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#transcript"><code>transcript</code></a> that
         overlaps with the CDS</dd></dl>
</li>
</ul>
<a name="liesInCDSIntron(de.charite.compbio.jannovar.reference.GenomePosition)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>liesInCDSIntron</h4>
<pre>public&nbsp;boolean&nbsp;liesInCDSIntron(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference">GenomePosition</a>&nbsp;pos)</pre>
<dl><dt><span class="strong">Returns:</span></dt><dd><code>true</code> if the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomePosition</code></a> lies within an intron of <a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html#transcript"><code>transcript</code></a> that
         overlaps with the CDS</dd></dl>
</li>
</ul>
<a name="overlapsWithTranslationalStartSite(de.charite.compbio.jannovar.reference.GenomeInterval)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>overlapsWithTranslationalStartSite</h4>
<pre>public&nbsp;boolean&nbsp;overlapsWithTranslationalStartSite(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</pre>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>interval</code> - the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> to use for querying</dd>
<dt><span class="strong">Returns:</span></dt><dd><code>true</code> if the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> overlaps with the translational start site</dd></dl>
</li>
</ul>
<a name="liesInTranslationalStartSite(de.charite.compbio.jannovar.reference.GenomePosition)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>liesInTranslationalStartSite</h4>
<pre>public&nbsp;boolean&nbsp;liesInTranslationalStartSite(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference">GenomePosition</a>&nbsp;pos)</pre>
<dl><dt><span class="strong">Returns:</span></dt><dd><code>true</code> if the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomePosition</code></a> lies within the translational start site</dd></dl>
</li>
</ul>
<a name="overlapsWithTranslationalStopSite(de.charite.compbio.jannovar.reference.GenomeInterval)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>overlapsWithTranslationalStopSite</h4>
<pre>public&nbsp;boolean&nbsp;overlapsWithTranslationalStopSite(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</pre>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>interval</code> - the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> to use for querying</dd>
<dt><span class="strong">Returns:</span></dt><dd><code>true</code> if the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> overlaps with the translational stop site</dd></dl>
</li>
</ul>
<a name="liesInTranslationalStopSite(de.charite.compbio.jannovar.reference.GenomePosition)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>liesInTranslationalStopSite</h4>
<pre>public&nbsp;boolean&nbsp;liesInTranslationalStopSite(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference">GenomePosition</a>&nbsp;pos)</pre>
<dl><dt><span class="strong">Returns:</span></dt><dd><code>true</code> if the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomePosition</code></a> lies within the translational stop site</dd></dl>
</li>
</ul>
<a name="overlapsWithSpliceRegion(de.charite.compbio.jannovar.reference.GenomeInterval)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>overlapsWithSpliceRegion</h4>
<pre>public&nbsp;boolean&nbsp;overlapsWithSpliceRegion(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</pre>
<div class="block">Returns whether the given <code>interval</code> overlaps with a splice region.

 A splice_region_variant is a sequence variant in which a change has occurred within the region of the splice
 site, either within 1-3 bases of the exon or 3-8 bases of the intron.</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>interval</code> - the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> to use for querying</dd>
<dt><span class="strong">Returns:</span></dt><dd><code>true</code> if the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> overlaps with a splice region.</dd></dl>
</li>
</ul>
<a name="liesInSpliceRegion(de.charite.compbio.jannovar.reference.GenomePosition)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>liesInSpliceRegion</h4>
<pre>public&nbsp;boolean&nbsp;liesInSpliceRegion(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference">GenomePosition</a>&nbsp;pos)</pre>
<div class="block">Returns whether the given <code>pos</code> lies within a splice region.

 A splice_region_variant is a sequence variant in which a change has occurred within the region of the splice
 site, either within 1-3 bases of the exon or 3-8 bases of the intron.</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>pos</code> - the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomePosition</code></a> to use for querying</dd>
<dt><span class="strong">Returns:</span></dt><dd><code>true</code> if the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomePosition</code></a> lies within a splice donor site.</dd></dl>
</li>
</ul>
<a name="overlapsWithSpliceDonorSite(de.charite.compbio.jannovar.reference.GenomeInterval)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>overlapsWithSpliceDonorSite</h4>
<pre>public&nbsp;boolean&nbsp;overlapsWithSpliceDonorSite(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</pre>
<div class="block">Returns whether the given <code>interval</code> overlaps with a splice donor site.

 A splice variant that changes the 2 base pair region at the 5' end of an intron.</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>interval</code> - the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> to use for querying</dd>
<dt><span class="strong">Returns:</span></dt><dd><code>true</code> if the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> overlaps with a splice donor site.</dd></dl>
</li>
</ul>
<a name="liesInSpliceDonorSite(de.charite.compbio.jannovar.reference.GenomePosition)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>liesInSpliceDonorSite</h4>
<pre>public&nbsp;boolean&nbsp;liesInSpliceDonorSite(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference">GenomePosition</a>&nbsp;pos)</pre>
<div class="block">Returns whether the given <code>pos</code> lies within a splice donor site.

 A splice variant that changes the 2 base pair region at the 5' end of an intron.</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>pos</code> - the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomePosition</code></a> to use for querying</dd>
<dt><span class="strong">Returns:</span></dt><dd><code>true</code> if the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomePosition</code></a> lies within a splice donor site.</dd></dl>
</li>
</ul>
<a name="overlapsWithSpliceAcceptorSite(de.charite.compbio.jannovar.reference.GenomeInterval)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>overlapsWithSpliceAcceptorSite</h4>
<pre>public&nbsp;boolean&nbsp;overlapsWithSpliceAcceptorSite(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</pre>
<div class="block">Returns whether the given <code>interval</code> overlaps with a splice acceptor site.

 A splice variant that changes the 2 base pair region at the 3' end of an intron.</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>interval</code> - the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> to use for querying</dd>
<dt><span class="strong">Returns:</span></dt><dd><code>true</code> if the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> overlaps with a splice acceptor site.</dd></dl>
</li>
</ul>
<a name="liesInSpliceAcceptorSite(de.charite.compbio.jannovar.reference.GenomePosition)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>liesInSpliceAcceptorSite</h4>
<pre>public&nbsp;boolean&nbsp;liesInSpliceAcceptorSite(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference">GenomePosition</a>&nbsp;pos)</pre>
<div class="block">Returns whether the given <code>pos</code> lies within a splice acceptor site.

 A splice variant that changes the 2 base pair region at the 3' end of an intron.</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>pos</code> - the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomePosition</code></a> to use for querying</dd>
<dt><span class="strong">Returns:</span></dt><dd><code>true</code> if the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomePosition</code></a> lies within a splice acceptor site.</dd></dl>
</li>
</ul>
<a name="getUpstreamInterval()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getUpstreamInterval</h4>
<pre>public&nbsp;<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;getUpstreamInterval()</pre>
<dl><dt><span class="strong">Returns:</span></dt><dd><a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> that gives the upstream region of the transcript.</dd></dl>
</li>
</ul>
<a name="overlapsWithUpstreamRegion(de.charite.compbio.jannovar.reference.GenomeInterval)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>overlapsWithUpstreamRegion</h4>
<pre>public&nbsp;boolean&nbsp;overlapsWithUpstreamRegion(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</pre>
<div class="block">Returns whether the given <code>interval</code> overlaps with the upstream region of the transcript.

 The upstream region of the transcript is up to 1000 bp upstream of the transcript.</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>interval</code> - the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> to use for querying</dd>
<dt><span class="strong">Returns:</span></dt><dd><code>true</code> if the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> overlaps with the upstream region of the transcript.</dd></dl>
</li>
</ul>
<a name="liesInUpstreamRegion(de.charite.compbio.jannovar.reference.GenomePosition)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>liesInUpstreamRegion</h4>
<pre>public&nbsp;boolean&nbsp;liesInUpstreamRegion(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference">GenomePosition</a>&nbsp;pos)</pre>
<div class="block">Returns whether the given <code>pos</code> lies within with the upstream region of the transcript.

 The upstream region of the transcript is up to 1000 bp upstream of the transcript.</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>pos</code> - the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomePosition</code></a> to use for querying</dd>
<dt><span class="strong">Returns:</span></dt><dd><code>true</code> if the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomePosition</code></a> lies within the upstream region of the transcript.</dd></dl>
</li>
</ul>
<a name="getDownstreamInterval()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getDownstreamInterval</h4>
<pre>public&nbsp;<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;getDownstreamInterval()</pre>
<dl><dt><span class="strong">Returns:</span></dt><dd><a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> that gives the downstream region of the transcript.</dd></dl>
</li>
</ul>
<a name="overlapsWithDownstreamRegion(de.charite.compbio.jannovar.reference.GenomeInterval)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>overlapsWithDownstreamRegion</h4>
<pre>public&nbsp;boolean&nbsp;overlapsWithDownstreamRegion(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</pre>
<div class="block">Returns whether the given <code>interval</code> overlaps with the downstream region of the transcript.

 The upstream region of the transcript is up to 1000 bp upstream of the transcript.</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>interval</code> - the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> to use for querying</dd>
<dt><span class="strong">Returns:</span></dt><dd><code>true</code> if the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> overlaps with the downstream region of the transcript.</dd></dl>
</li>
</ul>
<a name="liesInDownstreamRegion(de.charite.compbio.jannovar.reference.GenomePosition)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>liesInDownstreamRegion</h4>
<pre>public&nbsp;boolean&nbsp;liesInDownstreamRegion(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference">GenomePosition</a>&nbsp;pos)</pre>
<div class="block">Returns whether the given <code>pos</code> lies within with the downstream region of the transcript.

 The upstream region of the transcript is up to 1000 bp downstream of the transcript.</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>pos</code> - the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomePosition</code></a> to use for querying</dd>
<dt><span class="strong">Returns:</span></dt><dd><code>true</code> if the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomePosition</code></a> lies within the downstream region of the transcript.</dd></dl>
</li>
</ul>
<a name="overlapsWithFivePrimeUTR(de.charite.compbio.jannovar.reference.GenomeInterval)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>overlapsWithFivePrimeUTR</h4>
<pre>public&nbsp;boolean&nbsp;overlapsWithFivePrimeUTR(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</pre>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>interval</code> - the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> to use for querying</dd>
<dt><span class="strong">Returns:</span></dt><dd><code>true</code> if the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> overlaps with the 5' UTR</dd></dl>
</li>
</ul>
<a name="liesInFivePrimeUTR(de.charite.compbio.jannovar.reference.GenomePosition)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>liesInFivePrimeUTR</h4>
<pre>public&nbsp;boolean&nbsp;liesInFivePrimeUTR(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference">GenomePosition</a>&nbsp;pos)</pre>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>pos</code> - the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomePosition</code></a> to use for querying</dd>
<dt><span class="strong">Returns:</span></dt><dd><code>true</code> if the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomePosition</code></a> lies in the 5' UTR</dd></dl>
</li>
</ul>
<a name="overlapsWithThreePrimeUTR(de.charite.compbio.jannovar.reference.GenomeInterval)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>overlapsWithThreePrimeUTR</h4>
<pre>public&nbsp;boolean&nbsp;overlapsWithThreePrimeUTR(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</pre>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>interval</code> - the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> to use for querying</dd>
<dt><span class="strong">Returns:</span></dt><dd><code>true</code> if the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> overlaps with the 3' UTR</dd></dl>
</li>
</ul>
<a name="liesInThreePrimeUTR(de.charite.compbio.jannovar.reference.GenomePosition)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>liesInThreePrimeUTR</h4>
<pre>public&nbsp;boolean&nbsp;liesInThreePrimeUTR(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference">GenomePosition</a>&nbsp;pos)</pre>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>pos</code> - the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomePosition</code></a> to use for querying</dd>
<dt><span class="strong">Returns:</span></dt><dd><code>true</code> if the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomePosition</code></a> lies in the 3' UTR</dd></dl>
</li>
</ul>
<a name="liesInIntron(de.charite.compbio.jannovar.reference.GenomeInterval)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>liesInIntron</h4>
<pre>public&nbsp;boolean&nbsp;liesInIntron(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</pre>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>interval</code> - the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> to use for querying</dd>
<dt><span class="strong">Returns:</span></dt><dd><code>true</code> if the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> falls fully into an intron</dd></dl>
</li>
</ul>
<a name="liesInExon(de.charite.compbio.jannovar.reference.GenomeInterval)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>liesInExon</h4>
<pre>public&nbsp;boolean&nbsp;liesInExon(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</pre>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>interval</code> - the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> to use for querying</dd>
<dt><span class="strong">Returns:</span></dt><dd><code>true</code> if the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> falls fully into an exon</dd></dl>
</li>
</ul>
<a name="liesInExon(de.charite.compbio.jannovar.reference.GenomePosition)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>liesInExon</h4>
<pre>public&nbsp;boolean&nbsp;liesInExon(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference">GenomePosition</a>&nbsp;pos)</pre>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>pos</code> - the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomePosition</code></a> to use for querying</dd>
<dt><span class="strong">Returns:</span></dt><dd><code>true</code> if the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomePosition.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomePosition</code></a> points to a base an exon</dd></dl>
</li>
</ul>
<a name="overlapsWithExon(de.charite.compbio.jannovar.reference.GenomeInterval)">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>overlapsWithExon</h4>
<pre>public&nbsp;boolean&nbsp;overlapsWithExon(<a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference">GenomeInterval</a>&nbsp;interval)</pre>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>interval</code> - the <a href="../../../../../de/charite/compbio/jannovar/reference/GenomeInterval.html" title="class in de.charite.compbio.jannovar.reference"><code>GenomeInterval</code></a> to use for the overlap checking</dd>
<dt><span class="strong">Returns:</span></dt><dd><code>true</code> if the interval overlaps with an exon</dd></dl>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<!-- ========= END OF CLASS DATA ========= -->
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar_bottom">
<!--   -->
</a><a href="#skip-navbar_bottom" title="Skip navigation links"></a><a name="navbar_bottom_firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/TranscriptSequenceOntologyDecorator.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../index-all.html">Index</a></li>
<li><a href="../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSequenceDecorator.html" title="class in de.charite.compbio.jannovar.reference"><span class="strong">Prev Class</span></a></li>
<li><a href="../../../../../de/charite/compbio/jannovar/reference/TranscriptSupportLevels.html" title="interface in de.charite.compbio.jannovar.reference"><span class="strong">Next Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../../../../../index.html?de/charite/compbio/jannovar/reference/TranscriptSequenceOntologyDecorator.html" target="_top">Frames</a></li>
<li><a href="TranscriptSequenceOntologyDecorator.html" target="_top">No Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../../../../../allclasses-noframe.html">All Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor_summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor_detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_detail">Method</a></li>
</ul>
</div>
<a name="skip-navbar_bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
<p class="legalCopy"><small>Copyright &#169; 2015. All rights reserved.</small></p>
</body>
</html>
